// DEBUG/index.js
module.exports = function DEBUG (filename) {
  return function (sid) { return create_context(filename, sid) }
}

const DEFAULT_FLAGS = {
  show_default_entries: true
}

const scope = typeof window !== 'undefined' ? window : global

if (!scope.__DEBUG_GLOBAL_STATE__) {
  scope.__DEBUG_GLOBAL_STATE__ = {
    flags: { ...DEFAULT_FLAGS },
    listeners: []
  }
}

const state = scope.__DEBUG_GLOBAL_STATE__

function set_flag (name, value) {
  if (state.flags[name] === value) return
  state.flags[name] = value
  state.listeners.forEach(notify)

  function notify (listener) { listener(name, value) }
}

function get_flag (name) { return state.flags[name] }

function get_flags () { return { ...state.flags } }

function on_change (listener) {
  state.listeners.push(listener)
  return unsubscribe

  function unsubscribe () { state.listeners = state.listeners.filter(keep) }
  function keep (l) { return l !== listener }
}

function create_context (filename, sid) {
  return {
    get_flag,
    get_flags,
    set_flag,
    on_change,
    meta: { component: filename, sid }
  }
}
