// tile_manager/tile_manager.js
const STATE = require('STATE')
const statedb = STATE(__filename)
const { get } = statedb(fallback_module)
const net = require('net_helper')

const theme_widget = require('theme_widget')
const tab_group = require('tab_group')

module.exports = tile_manager

async function tile_manager (opts, invite) {
  const { id, sdb } = await get(opts.sid)
  const { drive } = sdb

  const on = {
    style: inject
  }

  const layout = {
    direction: null,
    tiles: [], // array of { id, element, sid }
    collapsed: false // responsive collapse state
  }

  const cached_actions = {
    update_actions_for_app: null,
    update_quick_actions_for_app: null,
    update_steps_wizard_for_app: null
  }

  let tile_counter = 0
  const COLLAPSE_THRESHOLD = 600
  const { io, _ } = net(id)

  const el = document.createElement('div')
  const shadow = el.attachShadow({ mode: 'closed' })
  shadow.innerHTML = `
  <div class="tile-manager">
    <div class="tile-slot" data-tile-id="0"></div>
  </div>
  `
  const sheet = new CSSStyleSheet()
  shadow.adoptedStyleSheets = [sheet]
  const container = shadow.querySelector('.tile-manager')

  const subs = await sdb.watch(onbatch)

  io.on.up = onmessage_from_root
  if (invite) io.accept(invite)

  const resize_observer = new ResizeObserver(entries => {
    for (const entry of entries) {
      const width = entry.contentRect.width
      handle_resize(width)
    }
  })
  resize_observer.observe(el)

  await create_tile(0)

  return el

  // ---------------------------
  // RESPONSIVE COLLAPSE
  // ---------------------------

  function handle_resize (width) {
    if (layout.tiles.length < 2) return

    const should_collapse = width < COLLAPSE_THRESHOLD

    if (should_collapse && !layout.collapsed) {
      collapse_split()
    } else if (!should_collapse && layout.collapsed) {
      expand_split()
    }
  }

  function collapse_split () {
    console.error('tile_manager: collapsing split')
    layout.collapsed = true

    for (const tile of layout.tiles) {
      if (tile.id !== 0 && tile.element) {
        tile.element.parentNode.style.display = 'none'
      }
    }

    container.classList.add('collapsed')
    container.classList.remove('horizontal', 'vertical')
  }

  function expand_split () {
    console.error('tile_manager: expanding split')
    layout.collapsed = false

    for (const tile of layout.tiles) {
      if (tile.id !== 0 && tile.element) {
        tile.element.parentNode.style.display = ''
      }
    }

    container.classList.remove('collapsed')
    if (layout.direction) {
      container.classList.add(layout.direction)
    }
  }

  // ---------------------------
  // TILE MANAGEMENT
  // ---------------------------

  async function create_tile (slot_index) {
    const tile_id = tile_counter++
    const tile_slot = container.querySelector(`[data-tile-id="${slot_index}"]`)

    if (!tile_slot) {
      console.error('tile_manager: slot not found for index', slot_index)
      return
    }

    const sub_entry = subs[tile_id] || { sid: opts.sid }

    const tile_info = {
      id: tile_id,
      element: null,
      slot: tile_slot,
      sid: sub_entry.sid
    }
    layout.tiles.push(tile_info)
    console.error('tile_manager: tile_info added for tile', tile_id)

    io.on[`tile_${tile_id}`] = create_tile_protocol(tile_id)
    console.error('tile_manager: handler registered for tile_' + tile_id)
    const tile_invite = io.invite(`tile_${tile_id}`, { up: id })
    console.error('tile_manager: invite created for tile_' + tile_id, '_[tile_' + tile_id + '] exists:', !!_[`tile_${tile_id}`])

    const component = tile_id === 0 ? theme_widget : tab_group
    const tile_el = await component(
      { ...sub_entry, ids: { up: id } },
      tile_invite
    )

    tile_info.element = tile_el
    tile_slot.appendChild(tile_el)

    console.error('tile_manager: created tile', tile_id)
    return tile_id
  }

  function create_tile_protocol (tile_id) {
    return function tile_protocol (msg) {
      const { type, data } = msg
      console.error(`tile_manager: message from tile_${tile_id}`, type, data)

      if (tile_id === 0 && type === 'link_tab_close_clicked') {
        console.error('tile_manager: link tab close clicked, merging tile', data.tile_id)
        handle_merge(data.tile_id)
        return
      }

      if (tile_id === 0 && type === 'link_tab_clicked') {
        console.error('tile_manager: link tab clicked, focusing tile', data.tile_id)
        focus_tile(data.tile_id)
        return
      }

      if (type === 'action_auto_completed' || type === 'action_complete') {
        console.error('tile_manager: action completed, checking for split', data)
        const action = data?.selected_action
        console.error('tile_manager: action name:', action?.name)
        if (action?.name === 'Split Tile') {
          let direction = null

          if (data.result) {
            try {
              const results = JSON.parse(data.result)
              direction = results[0] // First step's data is the direction
              console.error('tile_manager: direction from result:', direction)
            } catch (e) {
              console.error('tile_manager: failed to parse result', e)
            }
          }

          if (!direction) {
            const split_step = action.steps.find(s => s.component === 'form_tile_split_choice')
            console.error('tile_manager: split_step found:', split_step)
            if (split_step && split_step.data) {
              direction = split_step.data
            }
          }

          if (direction) {
            console.error('tile_manager: split requested', direction, 'from tile', tile_id)
            handle_split(tile_id, direction)
          } else {
            console.error('tile_manager: no direction found, cannot split')
          }
        }
      }

      if (_.up) {
        _.up(type, data, msg.head ? { cause: msg.head } : {})
      }
    }
  }

  async function handle_split (source_tile_id, direction) {
    if (layout.tiles.length >= 2) {
      console.error('tile_manager: already split, nested splitting not yet supported')
      return
    }

    const is_horizontal = direction === 'left' || direction === 'right'
    layout.direction = is_horizontal ? 'horizontal' : 'vertical'
    layout.split_direction = direction
    console.error('tile_manager: layout direction set to', layout.direction)

    container.classList.remove('horizontal', 'vertical')
    container.classList.add(layout.direction)

    const new_slot = document.createElement('div')
    new_slot.className = 'tile-slot'
    new_slot.setAttribute('data-tile-id', '1')

    if (direction === 'left' || direction === 'up') {
      container.insertBefore(new_slot, container.firstChild)
    } else {
      container.appendChild(new_slot)
    }
    console.error('tile_manager: new slot created')

    const new_tile_id = await create_tile(1)

    send_cached_actions_to_tile(1)

    const new_tile_send = _[`tile_${new_tile_id}`]
    if (new_tile_send) {
      console.error('tile_manager: sending create_default_tab to new tile')
      new_tile_send('create_default_tab', {
        name: 'New Tab',
        program: 'text_editor',
        tile_id: new_tile_id
      }, {})
    }

    const root_tile_send = _.tile_0
    if (root_tile_send) {
      console.error('tile_manager: sending add_link_tab to root tile')
      root_tile_send('add_link_tab', { tile_id: new_tile_id, direction, name: 'Split ' + direction }, {})
    }

    console.error('tile_manager: split complete', layout.direction, layout.tiles.length, 'tiles')
  }

  function handle_merge (tile_id_to_remove) {
    console.error('tile_manager: handle_merge', tile_id_to_remove)

    const tile_info = layout.tiles.find(t => t.id === tile_id_to_remove)
    if (!tile_info) {
      console.error('tile_manager: tile not found for merge', tile_id_to_remove)
      return
    }

    if (tile_info.element && tile_info.element.parentNode) {
      tile_info.element.parentNode.remove()
    }

    layout.tiles = layout.tiles.filter(t => t.id !== tile_id_to_remove)
    layout.direction = null
    layout.split_direction = null
    layout.collapsed = false

    container.classList.remove('horizontal', 'vertical', 'collapsed')

    const root_tile_send = _.tile_0
    if (root_tile_send) {
      console.error('tile_manager: sending remove_link_tab to root tile')
      root_tile_send('remove_link_tab', { tile_id: tile_id_to_remove }, {})
    }

    console.error('tile_manager: merge complete, tiles remaining:', layout.tiles.length)
  }

  function focus_tile (tile_id) {
    console.error('tile_manager: focus_tile', tile_id)

    if (layout.collapsed) {
      expand_split()
    }

    for (const tile of layout.tiles) {
      const slot = tile.element?.parentNode
      if (slot) {
        slot.classList.toggle('focused', tile.id === tile_id)
      }
    }

    const tile_info = layout.tiles.find(t => t.id === tile_id)
    if (tile_info?.element) {
      tile_info.element.scrollIntoView({ behavior: 'smooth', block: 'nearest' })
    }
  }

  function send_cached_actions_to_tile (tile_id) {
    const send = _[`tile_${tile_id}`]
    if (!send) {
      console.error('tile_manager: cannot send cached actions, tile_' + tile_id + ' not found')
      return
    }

    console.error('tile_manager: sending cached actions to tile_' + tile_id)

    for (const [type, data] of Object.entries(cached_actions)) {
      if (data !== null) {
        console.error('tile_manager: sending cached', type, 'to tile_' + tile_id)
        send(type, data, {})
      }
    }
  }

  // ---------------------------
  // MESSAGE HANDLERS
  // ---------------------------

  function onmessage_from_root (msg) {
    const { type, data } = msg
    console.error('tile_manager: message from root', type, 'tiles:', layout.tiles.length)

    // Cache action data for new tiles
    if (type === 'update_actions_for_app' || type === 'update_quick_actions_for_app' || type === 'update_steps_wizard_for_app') {
      cached_actions[type] = data
      console.error('tile_manager: cached', type)
    }

    // Forward to all tiles
    for (const tile of layout.tiles) {
      const send = _[`tile_${tile.id}`]
      console.error('tile_manager: forwarding to tile_' + tile.id, 'send exists:', !!send)
      if (send) {
        send(type, data, msg.head ? { cause: msg.head } : {})
      }
    }
  }

  // ---------------------------
  // BATCH HANDLER
  // ---------------------------

  async function onbatch (batch) {
    for (const { type, paths } of batch) {
      const data = await Promise.all(paths.map(load_path_raw))
      const func = on[type] || fail
      func({ data, type })
    }

    function load_path_raw (path) { return drive.get(path).then(read_drive_file_raw) }
    function read_drive_file_raw (file) { return file.raw }
  }

  function inject ({ data }) { sheet.replaceSync(data[0]) }

  function fail ({ data, type }) { console.warn('tile_manager: invalid message', { cause: { data, type } }) }
}

function fallback_module () {
  return {
    api: fallback_instance,
    _: {
      theme_widget: {
        $: ''
      },
      tab_group: {
        $: ''
      },
      net_helper: {
        $: ''
      }
    }
  }

  function fallback_instance () {
    return {
      _: {
        theme_widget: {
          0: '',
          mapping: {
            style: 'style',
            icons: 'icons',
            commands: 'commands',
            scroll: 'scroll',
            actions: 'actions',
            hardcons: 'hardcons',
            files: 'files',
            highlight: 'highlight',
            active_tab: 'active_tab',
            entries: 'entries',
            runtime: 'runtime',
            mode: 'mode',
            flags: 'flags',
            keybinds: 'keybinds',
            undo: 'undo',
            focused: 'focused',
            temp_actions: 'temp_actions',
            temp_quick_actions: 'temp_quick_actions',
            prefs: 'prefs',
            variables: 'variables',
            data: 'data',
            docs: 'docs',
            docs_style: 'docs_style'
          }
        },
        tab_group: {
          1: '',
          2: '',
          3: '',
          mapping: {
            style: 'style',
            icons: 'icons',
            actions: 'actions',
            docs: 'docs',
            docs_style: 'docs_style'
          }
        },
        net_helper: {
          0: ''
        }
      },
      drive: {
        'style/': {
          'tile_manager.css': {
            $ref: 'style/tile_manager.css'
          }
        }
      }
    }
  }
}
