// ui_gallery/index.js
const STATE = require('STATE')
const statedb = STATE(__filename)
const admin_api = statedb.admin()
const admin_on = {}
admin_api.on(handle_admin_message)
const { sdb, io: sdbio, id } = statedb(fallback_module)
const { drive, admin } = sdb
const net = require('net_helper')
const DOCS = require('DOCS')
const docs = DOCS(__filename)()
const docs_admin = docs.admin
/******************************************************************************
  PAGE
******************************************************************************/
const navbar = require('menu')
const theme_widget = require('theme_widget')
const taskbar = require('taskbar')
const tabsbar = require('tabsbar')
const action_bar = require('action_bar')
const program_container = require('program_container')
const tabs = require('tabs')
const console_history = require('console_history')
const actions = require('actions')
const tabbed_editor = require('tabbed_editor')
const task_manager = require('task_manager')
const quick_actions = require('quick_actions')
const graph_explorer_wrapper = require('graph_explorer_wrapper')
const editor = require('quick_editor')
const action_executor = require('action_executor')
const steps_wizard = require('steps_wizard')
const { resource } = require('helpers')

const imports = {
  theme_widget,
  taskbar,
  tabsbar,
  action_bar,
  program_container,
  tabs,
  console_history,
  actions,
  tabbed_editor,
  task_manager,
  quick_actions,
  graph_explorer_wrapper,
  action_executor,
  steps_wizard
}
module.exports = ui_gallery

/******************************************************************************
  PAGE BOOT
******************************************************************************/
async function ui_gallery (opts = {}) {
  // ----------------------------------------
  // ID + JSON STATE
  // ----------------------------------------
  let resize_enabled = true
  const on = {
    style: inject,
    resize_container: update_resize,
    ...sdb.admin.status.dataset.drive,
    ...sdb.admin
  }
  // const status = {}
  // ----------------------------------------
  // TEMPLATE
  // ----------------------------------------
  const el = document.createElement('div')
  const shadow = el.attachShadow({ mode: 'closed' })
  shadow.innerHTML = `
  <div class="navbar-slot"></div>
  <div class="components-wrapper-container">
    <div class="components-wrapper"></div>
  </div>`
  document.body.style.margin = 0
  document.body.style.backgroundColor = '#d8dee9'

  // ----------------------------------------
  // ELEMENTS
  // ----------------------------------------

  const navbar_slot = shadow.querySelector('.navbar-slot')
  const components_wrapper = shadow.querySelector('.components-wrapper')
  const sheet = new CSSStyleSheet()
  shadow.adoptedStyleSheets = [sheet]

  const entries = Object.entries(imports)
  const wrappers = []
  const names = entries.map(get_entry_name)
  let current_selected_wrapper = null

  function get_entry_name (entry) { return entry[0] }

  const url_params = new URLSearchParams(window.location.search)
  const checked_param = url_params.get('checked')
  const selected_name_param = url_params.get('selected')
  let initial_checked_indices = []

  if (checked_param) {
    try {
      const parsed = JSON.parse(checked_param)
      if (Array.isArray(parsed) && parsed.every(Number.isInteger)) {
        initial_checked_indices = parsed
      } else {
        console.warn('Invalid "checked" URL parameter format.')
      }
    } catch (e) {
      console.error('Error parsing "checked" URL parameter:', e)
    }
  }

  const menu_callbacks = {
    on_checkbox_change: handle_checkbox_change,
    on_label_click: handle_label_click,
    on_select_all_toggle: handle_select_all_toggle,
    on_resize_toggle: handle_resize_toggle
  }
  const item = resource()
  sdbio.on(register_io_port)
  const { io: theme_widget_io, _: theme_widget_send } = net(id)
  theme_widget_io.on.theme_widget = theme_widget_protocol

  // Create io handles for all preview components to provide up-channel invites
  const preview_ios = {}
  const preview_names = Object.keys(imports)
  for (const name of preview_names) {
    if (name === 'theme_widget') continue
    const { io } = net(id)
    // Register minimal protocol handler so io.invite() can work
    io.on[name] = () => {}
    io.on.up = () => {}
    preview_ios[name] = io
  }

  function register_io_port (port) {
    const { by, to } = port
    item.set(port.to, port)

    port.onmessage = on_port_message

    function on_port_message (event) {
      const txt = event.data
      const key = `[${by} -> ${to}]`
      console.log('[ port-stuff ]', key)

      on[txt.type](...txt.data)
    }
  }

  const editor_subs = await sdb.get_sub('ui_gallery>quick_editor')
  // const subs = await sdb.watch(onbatch)
  const subs = (await sdb.watch(onbatch)).filter(is_even_index)

  function is_even_index (_, index) { return index % 2 === 0 }

  console.log('Page subs', subs)
  const nav_menu_element = await navbar(subs[names.length], names, initial_checked_indices, menu_callbacks)

  const main_editor = editor_subs[0] ? await editor(editor_subs[0]) : null
  navbar_slot.replaceWith(nav_menu_element, main_editor || document.createElement('div'))
  await create_component(entries)
  update_resize(resize_enabled)
  window.onload = scroll_to_initial_selected
  send_quick_editor_data()
  admin_on.import = send_quick_editor_data

  function theme_widget_protocol (msg) {
    const action_handlers = {
      set_docs_mode: handle_set_docs_mode,
      set_doc_display_handler: handle_set_doc_display_handler,
      focused_app_changed: handle_focused_app_changed
    }
    const handler = action_handlers[msg.type] || handle_fail
    handler(msg)

    function handle_set_docs_mode (msg) { docs_admin.set_docs_mode(msg.data.active) }
    function handle_set_doc_display_handler (msg) { docs_admin.set_doc_display_handler(msg.data.callback) }
    function handle_fail (msg) { console.warn('page: unhandled message from theme_widget', msg) }

    function handle_focused_app_changed (msg) {
      const actions = docs_admin.get_actions(msg.data.sid)
      update_actions_for_app(actions, msg)
    }

    async function update_actions_for_app (data, msg) {
      const focused_app = msg.data.type
      let actions_data = null
      let quick_actions_data = null
      let steps_wizard_data = null

      if (focused_app) {
        const component_actions = await get_component_actions(data)
        actions_data = component_actions.actions
        quick_actions_data = component_actions.quick_actions
        steps_wizard_data = component_actions.steps_wizard
      }

      const refs = msg.head ? { cause: msg.head } : {}
      theme_widget_send.theme_widget('update_actions_for_app', actions_data, refs)
      theme_widget_send.theme_widget('update_quick_actions_for_app', quick_actions_data, refs)
      theme_widget_send.theme_widget('update_steps_wizard_for_app', steps_wizard_data, refs)
    }

    async function get_component_actions (data) {
      const result_actions = []
      const result_quick_actions = []

      data.forEach(add_action_entry)

      return {
        actions: result_actions,
        quick_actions: result_quick_actions,
        steps_wizard: data
      }

      function add_action_entry (entry) {
        result_actions.push({
          action: entry.name,
          icon: entry.icon,
          pinned: entry.status.pinned,
          default: entry.status.default
        })
        result_quick_actions.push({
          name: entry.name,
          icon: entry.icon,
          total_steps: entry.steps.length
        })
      }
    }
  }
  return el
  async function create_component (entries_obj) {
    let index = 0
    for (const [name, factory] of entries_obj) {
      const is_initially_checked = initial_checked_indices.length === 0 || initial_checked_indices.includes(index + 1)
      const outer = document.createElement('div')
      outer.className = 'component-outer-wrapper'
      outer.style.display = is_initially_checked ? 'block' : 'none'
      outer.innerHTML = `
      <div class="component-name-label">${name}</div>
      <div class="component-wrapper"></div>
    `
      const inner = outer.querySelector('.component-wrapper')
      let component_content
      if (name === 'theme_widget') {
        component_content = await factory({ ...subs[index] }, theme_widget_io.invite('theme_widget', { up: id }))
      } else {
        const component_io = preview_ios[name]
        component_content = await factory({ ...subs[index] }, component_io.invite(name, { up: id }))
      }
      component_content.className = 'component-content'

      const node_id = admin.status.s2i[subs[index].sid]
      const editor_index = index + 1
      const component_editor = editor_subs[editor_index] ? await editor(editor_subs[editor_index]) : null
      inner.append(component_content, component_editor || document.createElement('div'))

      const result = {}
      const drive = admin.status.dataset.drive

      const modulepath = node_id.split(':')[0]
      const fields = admin.status.db.read_all(['state', modulepath])
      const nodes = Object.keys(fields).filter(is_state_node)

      function is_state_node (field) { return !isNaN(Number(field.split(':').at(-1))) }

      for (const node of nodes) {
        result[node] = {}
        const datasets = drive.list('', node)
        for (const dataset of datasets) {
          result[node][dataset] = {}
          const files = drive.list(dataset, node)
          for (const file of files) {
            result[node][dataset][file] = (await drive.get(dataset + file, node)).raw
          }
        }
      }

      if (editor_subs[editor_index]) {
        const editor_id = admin.status.a2i[admin.status.s2i[editor_subs[editor_index].sid]]
        const port = await item.get(editor_id)
        // await sdbio.at(editor_id)
        port.postMessage(result)
      }

      components_wrapper.appendChild(outer)
      wrappers[index] = { outer, inner, name, checkbox_state: is_initially_checked }
      index++
    }
  }

  function scroll_to_initial_selected () {
    if (selected_name_param) {
      const index = names.indexOf(selected_name_param)
      if (index !== -1 && wrappers[index]) {
        const target_wrapper = wrappers[index].outer
        if (target_wrapper.style.display !== 'none') {
          setTimeout(scroll_to_selected_wrapper, 100)

          function scroll_to_selected_wrapper () {
            target_wrapper.scrollIntoView({ behavior: 'auto', block: 'center' })
            clear_selection_highlight()
            target_wrapper.style.backgroundColor = '#2e3440'
            current_selected_wrapper = target_wrapper
          }
        }
      }
    }
  }

  function clear_selection_highlight () {
    if (current_selected_wrapper) {
      current_selected_wrapper.style.backgroundColor = ''
    }
    current_selected_wrapper = null
  }

  function update_url (selected_name = url_params.get('selected')) {
    const checked_indices = wrappers.reduce(collect_checked_index, [])

    function collect_checked_index (acc, wrapper_entry, index) {
      if (wrapper_entry.checkbox_state) { acc.push(index + 1) }
      return acc
    }

    const params = new URLSearchParams()
    if (checked_indices.length > 0 && checked_indices.length < wrappers.length) {
      params.set('checked', JSON.stringify(checked_indices))
    }
    const selected_index = names.indexOf(selected_name)
    if (selected_name && selected_index !== -1 && wrappers[selected_index].checkbox_state) {
      params.set('selected', selected_name)
    }
    const new_url = `${window.location.pathname}${params.toString() ? '?' + params.toString() : ''}`
    window.history.replaceState(null, '', new_url)
  }

  function handle_checkbox_change (detail) {
    const { index, checked } = detail
    if (wrappers[index]) {
      wrappers[index].outer.style.display = checked ? 'block' : 'none'
      wrappers[index].checkbox_state = checked
      update_url()
      if (!checked && current_selected_wrapper === wrappers[index].outer) {
        clear_selection_highlight()
        update_url(null)
      }
    }
  }

  function handle_label_click (detail) {
    const { index, name } = detail
    if (wrappers[index]) {
      const target_wrapper = wrappers[index].outer
      if (target_wrapper.style.display === 'none') {
        target_wrapper.style.display = 'block'
        wrappers[index].checkbox_state = true
      }
      target_wrapper.scrollIntoView({ behavior: 'smooth', block: 'center' })
      clear_selection_highlight()
      target_wrapper.style.backgroundColor = 'lightblue'
      current_selected_wrapper = target_wrapper
      update_url(name)
    }
  }

  function handle_select_all_toggle (detail) {
    const { selectAll: select_all } = detail
    wrappers.forEach(update_wrapper_visibility)

    function update_wrapper_visibility (wrapper_entry) {
      wrapper_entry.outer.style.display = select_all ? 'block' : 'none'
      wrapper_entry.checkbox_state = select_all
    }

    clear_selection_highlight()
    update_url(null)
  }

  function handle_resize_toggle () {
    console.log('handle_resize_toggle', resize_enabled)
    resize_enabled = !resize_enabled
    drive.put('resize_container/state.json', resize_enabled)
  }

  async function onbatch (batch) {
    for (const { type, paths } of batch) {
      const data = await Promise.all(paths.map(load_path_raw))
      const func = on[type] || fail
      func(data, type)
    }

    function load_path_raw (path) { return drive.get(path).then(read_drive_file_raw) }
    function read_drive_file_raw (file) { return file.raw }
  }
  function fail (data, type) { console.warn(__filename + 'invalid message', { cause: { data, type } }) }
  function inject (data) { sheet.replaceSync(data[0]) }
  function update_resize (data) {
    console.log('[ update_resize ]', data)
    resize_enabled = data
    wrappers.forEach(update_wrapper_resize)

    function update_wrapper_resize (wrap) {
      const wrapper = wrap.outer.querySelector('.component-wrapper')
      if (wrapper) {
        wrapper.style.resize = resize_enabled ? 'both' : 'none'
        wrapper.style.overflow = resize_enabled ? 'auto' : 'visible'
      }
    }
  }
  async function send_quick_editor_data () {
    const roots = admin.status.db.read(['root_datasets'])
    const result = {}
    roots.forEach(add_root_dataset)

    function add_root_dataset (root_dataset) {
      const root = root_dataset.name
      result[root] = {}
      const inputs = sdb.admin.get_dataset({ root }) || []
      inputs.forEach(add_input_type)

      function add_input_type (type) {
        result[root][type] = {}
        const datasets = sdb.admin.get_dataset({ root, type })

        if (!datasets) return
        Object.values(datasets).forEach(add_dataset_name)

        function add_dataset_name (dataset_name) {
          result[root][type][dataset_name] = {}
          const dataset_ids = sdb.admin.get_dataset({ root, type, name: dataset_name })
          dataset_ids.forEach(add_dataset_id)

          function add_dataset_id (dataset_id) {
            const files = admin.status.db.read([root, dataset_id]).files || []
            result[root][type][dataset_name][dataset_id] = {}
            files.forEach(add_file_data)

            function add_file_data (file_id) { result[root][type][dataset_name][dataset_id][file_id] = admin.status.db.read([root, file_id]) }
          }
        }
      }
    }

    if (!editor_subs[0]) return
    const editor_id = admin.status.a2i[admin.status.s2i[editor_subs[0].sid]]
    const port = await item.get(editor_id)
    // await sdbio.at(editor_id)
    port.postMessage(result)
  }
}
function fallback_module () {
  const menuname = 'menu'
  const names = [
    'theme_widget',
    'taskbar',
    'tabsbar',
    'action_bar',
    'program_container',
    'tabs',
    'console_history',
    'actions',
    'tabbed_editor',
    'task_manager',
    'quick_actions',
    'graph_explorer_wrapper',
    'action_executor',
    'steps_wizard'
  ]
  const subs = {}
  names.forEach(subgen)
  subs.helpers = 0
  subs.DOCS = 0
  subs.net_helper = 0
  subs.taskbar = {
    $: '',
    0: '',
    mapping: {
      style: 'style',
      icons: 'icons',
      actions: 'actions',
      prefs: 'prefs',
      variables: 'variables',
      data: 'data',
      hardcons: 'hardcons',
      docs: 'docs'
    }
  }
  subs.tabs = {
    $: '',
    0: '',
    mapping: {
      icons: 'icons',
      variables: 'variables',
      scroll: 'scroll',
      style: 'style',
      docs: 'docs',
      actions: 'actions'
    }
  }
  subs.program_container = {
    $: '',
    0: '',
    mapping: {
      style: 'style',
      flags: 'flags',
      commands: 'commands',
      icons: 'icons',
      scroll: 'scroll',
      actions: 'actions',
      hardcons: 'hardcons',
      files: 'files',
      highlight: 'highlight',
      active_tab: 'active_tab',
      entries: 'entries',
      runtime: 'runtime',
      mode: 'mode',
      keybinds: 'keybinds',
      undo: 'undo',
      docs_style: 'docs_style',
      docs: 'docs'
    }
  }
  subs.action_executor = {
    $: '',
    0: '',
    mapping: {
      style: 'style',
      variables: 'variables',
      docs: 'docs',
      data: 'data'
    }
  }
  subs.steps_wizard = {
    $: '',
    0: '',
    mapping: {
      style: 'style',
      docs: 'docs'
    }
  }
  subs.tabsbar = {
    $: '',
    0: '',
    mapping: {
      icons: 'icons',
      style: 'style',
      docs: 'docs',
      actions: 'actions'
    }
  }
  subs.action_bar = {
    $: '',
    0: '',
    mapping: {
      icons: 'icons',
      style: 'style',
      actions: 'actions',
      variables: 'variables',
      hardcons: 'hardcons',
      prefs: 'prefs',
      docs: 'docs'
    }
  }
  subs.console_history = {
    $: '',
    0: '',
    mapping: {
      style: 'style',
      commands: 'commands',
      icons: 'icons',
      scroll: 'scroll',
      docs: 'docs',
      actions: 'actions'
    }
  }
  subs.actions = {
    $: '',
    0: '',
    mapping: {
      actions: 'actions',
      icons: 'icons',
      hardcons: 'hardcons',
      style: 'style',
      docs: 'docs'
    }
  }
  subs.tabbed_editor = {
    $: '',
    0: '',
    mapping: {
      style: 'style',
      files: 'files',
      highlight: 'highlight',
      active_tab: 'active_tab',
      docs: 'docs'
    }
  }
  subs.task_manager = {
    $: '',
    0: '',
    mapping: {
      style: 'style',
      count: 'count',
      docs: 'docs',
      actions: 'actions'
    }
  }
  subs.quick_actions = {
    $: '',
    0: '',
    mapping: {
      style: 'style',
      icons: 'icons',
      actions: 'actions',
      hardcons: 'hardcons',
      prefs: 'prefs',
      docs: 'docs'
    }
  }
  subs[menuname] = {
    $: '',
    0: '',
    mapping: {
      style: 'style',
      docs: 'docs'
    }
  }
  subs.quick_editor = {
    $: '',
    mapping: {
      style: 'style',
      docs: 'docs'
    }
  }
  subs.theme_widget = {
    $: '',
    0: '',
    mapping: {
      style: 'style',
      commands: 'commands',
      icons: 'icons',
      scroll: 'scroll',
      actions: 'actions',
      hardcons: 'hardcons',
      files: 'files',
      highlight: 'highlight',
      active_tab: 'active_tab',
      entries: 'entries',
      runtime: 'runtime',
      mode: 'mode',
      flags: 'flags',
      keybinds: 'keybinds',
      undo: 'undo',
      focused: 'focused',
      temp_actions: 'temp_actions',
      temp_quick_actions: 'temp_quick_actions',
      prefs: 'prefs',
      variables: 'variables',
      data: 'data',
      docs_style: 'docs_style',
      docs: 'docs'
    }
  }
  subs.graph_explorer_wrapper = {
    $: '',
    0: '',
    mapping: {
      theme: 'style',
      entries: 'entries',
      runtime: 'runtime',
      mode: 'mode',
      flags: 'flags',
      keybinds: 'keybinds',
      undo: 'undo',
      docs: 'docs'
    }
  }
  for (let i = 0; i < Object.keys(subs).length - 1; i++) {
    subs.quick_editor[i] = quick_editor$
  }

  return {
    _: subs,
    drive: {
      'style/': {
        'theme.css': {
          raw: `
          .components-wrapper-container {
            padding-top: 10px; /* Adjust as needed */
          }

          .component-outer-wrapper {
            margin-bottom: 20px;
            padding: 0px 0px 10px 0px;
            transition: background-color 0.3s ease;
          }

          .component-name-label {
            background-color:transparent;
            padding: 8px 15px;
            text-align: center;
            font-weight: bold;
            color: #333;
          }

          .component-wrapper {
            width: 95%;
            margin: 0 auto;
            position: relative;
            padding: 15px;
            border: 3px solid #666;
            resize: none;
            overflow: visible;
            border-radius: 0px;
            background-color: #eceff4;
            min-height: 50px;
          }
          .component-content {
            width: 100%;
            height: 100%;
          }
          .toggle-switch {
            position: relative;
            display: inline-block;
            width: 50px;
            height: 26px;
          }

          .toggle-switch input {
            opacity: 0;
            width: 0;
            height: 0;
          }

          .slider {
            position: absolute;
            cursor: pointer;
            inset: 0;
            background-color: #ccc;
            border-radius: 26px;
            transition: 0.4s;
          }

          .slider::before {
            content: "";
            position: absolute;
            height: 20px;
            width: 20px;
            left: 3px;
            bottom: 3px;
            background-color: white;
            border-radius: 50%;
            transition: 0.4s;
          }

          input:checked + .slider {
            background-color: #2196F3;
          }

          input:checked + .slider::before {
            transform: translateX(24px);
          }
          .component-wrapper:hover::before {
            content: '';
            position: absolute;
            width: 100%;
            height: 100%;
            top: 0;
            left: 0;
            border: 4px solid skyblue;
            pointer-events: none;
            z-index: 15;
            resize: both;
            overflow: auto;
          }
          .quick-editor {
            position: absolute;
            z-index: 100;
            top: 0;
            right: 0;
          }
          .component-wrapper:hover .quick-editor {
            display: block;
          }
          .component-wrapper > .quick-editor {
            display: none;
            top: -5px;
            right: -10px;
          }`
        }
      },
      'resize_container/': {
        'state.json': {
          raw: 'false'
        }
      },
      'icons/': {},
      'variables/': {},
      'scroll/': {},
      'commands/': {},
      'actions/': {},
      'hardcons/': {},
      'files/': {},
      'highlight/': {},
      'count/': {},
      'entries/': {},
      'active_tab/': {},
      'runtime/': {},
      'mode/': {},
      'data/': {},
      'flags/': {},
      'keybinds/': {},
      'undo/': {},
      'focused/': {},
      'temp_actions/': {},
      'temp_quick_actions/': {},
      'prefs/': {},
      'docs_style/': {},
      'docs/': {}
    }
  }
  function quick_editor$ (args, tools, [quick_editor]) {
    const state = quick_editor()
    state.net = {
      page: {}
    }
    return state
  }
  function subgen (name) {
    subs[name] = {
      $: '',
      0: '',
      mapping: {
        style: 'style',
        docs: 'docs'
      }
    }
  }
}

function handle_admin_message (msg) {
  const { type } = msg
  admin_on[type] && admin_on[type]()
}
