// graph_explorer_wrapper/graph_explorer_wrapper.js
const STATE = require('STATE')
const statedb = STATE(__filename)
const { get } = statedb(fallback_module)
const net = require('net_helper')
const graph_explorer = require('graph-explorer')
const graphdb = require('./graphdb')

module.exports = graph_explorer_wrapper

async function graph_explorer_wrapper (opts, invite) {
  const { id, sdb } = await get(opts.sid)
  const { drive } = sdb

  let db = null
  let latest_entries = null
  let graph_explorer_connected = false

  // Protocol
  const { io, _ } = net(id)
  io.on = {
    up: io_up(),
    graph_explorer: io_graph_explorer()
  }
  if (invite) io.accept(invite)

  const on = {
    theme: inject,
    entries: on_entries
  }
  const db_handler = {
    db_get: params => db.get(params.path),
    db_has: params => db.has(params.path),
    db_is_empty: () => db.is_empty(),
    db_root: () => db.root(),
    db_keys: () => db.keys(),
    db_raw: () => db.raw()
  }
  const el = document.createElement('div')
  const shadow = el.attachShadow({ mode: 'closed' })

  const sheet = new CSSStyleSheet()
  shadow.adoptedStyleSheets = [sheet]

  const subs = await sdb.watch(onbatch)

  const explorer_el = await graph_explorer(subs[0], io.invite('graph_explorer', { storage: id }))
  graph_explorer_connected = true
  if (latest_entries) _.graph_explorer('db_initialized', {}, { entries: latest_entries })
  shadow.append(explorer_el)

  return el

  function io_up () {
    const parent_handlers = {
      execute_step: parent_execute_step,
      set_mode: parent_forward_to_graph_explorer,
      set_search_query: parent_forward_to_graph_explorer,
      select_nodes: parent_forward_to_graph_explorer,
      expand_node: parent_forward_to_graph_explorer,
      collapse_node: parent_forward_to_graph_explorer,
      toggle_node: parent_forward_to_graph_explorer,
      get_selected: parent_forward_to_graph_explorer,
      get_confirmed: parent_forward_to_graph_explorer,
      clear_selection: parent_forward_to_graph_explorer,
      set_flag: parent_forward_to_graph_explorer,
      scroll_to_node: parent_forward_to_graph_explorer,
      docs_toggle: parent_forward_to_graph_explorer
    }
    return function onmessage (msg) {
      const handler = parent_handlers[msg.type] || fail
      handler(msg)
    }
  }

  function parent_execute_step (msg) {
    const commands = get_step_commands(msg.data)
    for (const command of commands) {
      const refs = msg.head ? { cause: msg.head } : {}
      const data = command.data !== undefined ? command.data : {}
      _.graph_explorer(command.type, refs, data)
    }
  }

  function parent_forward_to_graph_explorer (msg) {
    const refs = msg.head ? { cause: msg.head } : {}
    _.graph_explorer(msg.type, refs, msg.data)
  }

  function get_step_commands (data) {
    if (!data) return []
    if (Array.isArray(data.commands)) return data.commands.filter(has_command_type)
    if (data.command && has_command_type(data.command)) return [data.command]
    if (has_command_type(data)) {
      return [{ type: data.type, data: data.data !== undefined ? data.data : {} }]
    }
    return []

    function has_command_type (command) {
      return command && typeof command.type === 'string' && command.type.length > 0
    }
  }

  async function onbatch (batch) {
    for (const { type, paths } of batch) {
      const data = await Promise.all(paths.map(load_path_raw))
      const handler = on[type] || fail
      handler({ data, type })
    }

    function load_path_raw (path) { return drive.get(path).then(read_drive_file_raw) }
    function read_drive_file_raw (file) { return file.raw }
  }

  function fail ({ data, type }) { console.warn('invalid message', { cause: { data, type } }) }
  function inject ({ data }) { sheet.replaceSync(data.join('\n')) }

  function on_entries ({ data }) {
    if (!data || !data[0]) {
      console.error('Entries data is missing or empty.')
      latest_entries = {}
      db = graphdb({})
      if (graph_explorer_connected) _.graph_explorer('db_initialized', {}, { entries: {} })
      return
    }

    let parsed_data
    try {
      parsed_data = typeof data[0] === 'string' ? JSON.parse(data[0]) : data[0]
    } catch (e) {
      console.error('Failed to parse entries data:', e)
      parsed_data = {}
    }

    if (typeof parsed_data !== 'object' || !parsed_data) {
      console.error('Parsed entries data is not a valid object.')
      parsed_data = {}
    }

    db = graphdb(parsed_data)
    latest_entries = parsed_data
    if (graph_explorer_connected) _.graph_explorer('db_initialized', {}, { entries: parsed_data })
  }

  // ---------------------------------------------------------
  // PROTOCOL
  // ---------------------------------------------------------

  function io_graph_explorer () {
    return function graph_explorer_protocol (msg) {
      const { type } = msg

      if (type.startsWith('db_')) {
        handle_db_request(msg)
      } else {
        _.up(msg.type, msg.head ? { cause: msg.head } : {}, msg.data)
      }
    }
  }

  function handle_db_request ({ head, type, data }) {
    const handler = db ? db_handler[type] || db_fail : db_fail
    _.graph_explorer('db_response', { cause: head }, { result: handler(data) })

    function db_fail () {
      const msg = db ? '[graph_explorer_wrapper] Unknown db operation:' : '[graph_explorer_wrapper] Database not initialized yet'
      console.warn(msg, type)
      return null
    }
  }
}

function fallback_module () {
  return {
    _: {
      'graph-explorer': {
        $: ''
      },
      './graphdb': {
        $: ''
      },
      net_helper: {
        $: ''
      }
    },
    api: fallback_instance
  }

  function fallback_instance () {
    return {
      _: {
        'graph-explorer': {
          $: '',
          0: '',
          mapping: {
            style: 'theme',
            runtime: 'runtime',
            mode: 'mode',
            flags: 'flags',
            keybinds: 'keybinds',
            undo: 'undo',
            docs: 'docs'
          }
        },
        './graphdb': {
          $: ''
        },
        net_helper: {
          0: ''
        }
      },
      drive: {
        'theme/': {
          'style.css': {
            raw: `
              :host {
              display: block;
              height: 100%;
              width: 100%;
              }
            `
          }
        },
        'entries/': {
          'entries.json': {
            $ref: 'entries.json'
          }
        },
        'runtime/': {
          'node_height.json': { raw: '16' },
          'vertical_scroll_value.json': { raw: '0' },
          'horizontal_scroll_value.json': { raw: '0' },
          'selected_instance_paths.json': { raw: '[]' },
          'confirmed_selected.json': { raw: '[]' },
          'instance_states.json': { raw: '{}' },
          'search_entry_states.json': { raw: '{}' },
          'last_clicked_node.json': { raw: 'null' },
          'view_order_tracking.json': { raw: '{}' }
        },
        'mode/': {
          'current_mode.json': { raw: '"menubar"' },
          'previous_mode.json': { raw: '"menubar"' },
          'search_query.json': { raw: '""' },
          'multi_select_enabled.json': { raw: 'false' },
          'select_between_enabled.json': { raw: 'false' }
        },
        'flags/': {
          'hubs.json': { raw: '"default"' },
          'selection.json': { raw: 'true' },
          'recursive_collapse.json': { raw: 'true' }
        },
        'keybinds/': {
          'navigation.json': {
            raw: JSON.stringify({
              ArrowUp: 'navigate_up_current_node',
              ArrowDown: 'navigate_down_current_node',
              'Control+ArrowDown': 'toggle_subs_for_current_node',
              'Control+ArrowUp': 'toggle_hubs_for_current_node',
              'Alt+s': 'multiselect_current_node',
              'Alt+b': 'select_between_current_node',
              'Control+m': 'toggle_search_mode',
              'Alt+j': 'jump_to_next_duplicate'
            })
          }
        },
        'undo/': {
          'stack.json': { raw: '[]' }
        },
        'docs/': {
          'README.md': {
            $ref: 'README.md'
          }
        }
      }
    }
  }
}
