// action_executor/action_executor.js
const STATE = require('STATE')
const statedb = STATE(__filename)
const { get } = statedb(fallback_module)
const net = require('net_helper')

const program = require('program')
const steps_wizard = require('steps_wizard')

const { form_input, input_test, form_tile_split_choice } = program

const component_modules = {
  form_input,
  input_test,
  form_tile_split_choice
  // Add more form input components here if needed
}

module.exports = action_executor

async function action_executor (opts, invite) {
  const { id, sdb } = await get(opts.sid)
  const { drive } = sdb
  const ids = opts.ids
  if (!ids || !ids.up) {
    throw new Error(`Component ${__filename} requires ids.up to be provided`)
  }

  const on = {
    style: inject
  }

  const el = document.createElement('div')
  const shadow = el.attachShadow({ mode: 'closed' })
  shadow.innerHTML = `
  <div class="container main">
    <program></program>
    <form-input></form-input>
    <steps-wizard></steps-wizard>
  </div>`

  const sheet = new CSSStyleSheet()
  shadow.adoptedStyleSheets = [sheet]
  const program_placeholder = shadow.querySelector('program')
  const form_input_placeholder = shadow.querySelector('form-input')
  const steps_wizard_placeholder = shadow.querySelector('steps-wizard')
  const { io, _ } = net(id)

  const subs = await sdb.watch(onbatch)

  let all_data = null
  let variables = []
  let selected_action = null

  io.on.up = onmessage
  io.on.program = program_protocol
  io.on.steps_wizard = steps_wizard_protocol

  // dynamic form input component SIDs
  for (const [index, [component_name]] of Object.entries(component_modules).entries()) {
    const final_index = index + 2
    io.on[component_name] = form_input_protocol(component_name)
  }
  if (invite) io.accept(invite)

  const program_el = await program({ ...subs[0], ids: { up: id } }, io.invite('program', { up: id }))
  program_el.classList.add('program-bar', 'hide')
  program_placeholder.replaceWith(program_el)

  const steps_wizard_el = await steps_wizard({ ...subs[1], ids: { up: id } }, io.invite('steps_wizard', { up: id }))
  steps_wizard_el.classList.add('steps-wizard-bar', 'hide')
  steps_wizard_placeholder.replaceWith(steps_wizard_el)

  const form_input_elements = {}

  for (const [index, [component_name, component_fn]] of Object.entries(component_modules).entries()) {
    const final_index = index + 2
    const sub_entry = subs[final_index] || { sid: opts.sid }
    const element = await component_fn({ ...sub_entry, ids: { up: id } }, io.invite(component_name, { up: id }))
    element.classList.add('form-inputs', 'hide')
    form_input_elements[component_name] = element
    form_input_placeholder.parentNode.insertBefore(element, form_input_placeholder)
  }

  form_input_placeholder.remove()

  const parent_handler = {
    update_steps_wizard_for_app,
    load_actions,
    action_submitted,
    update_data,
    activate_steps_wizard,
    form_data,
    render_form,
    selected_action: parent_selected_action,
    clean_up
  }

  return el

  async function onbatch (batch) {
    for (const { type, paths } of batch) {
      const data = await Promise.all(paths.map(load_path_raw))
      const func = on[type] || fail
      func(data, type)
    }

    function load_path_raw (path) { return drive.get(path).then(read_drive_file_raw) }
    function read_drive_file_raw (file) { return file.raw }
  }
  function fail (data, type) { console.warn('Unknown message type:', type, data) }
  function inject (data) { sheet.replaceSync(data[0]) }

  // --- Toggle Views ---
  function toggle_view (el, show) { el.classList.toggle('hide', !show) }
  function steps_toggle_view (display) { toggle_view(steps_wizard_el, display === 'block') }

  function render_form_component (component_name) {
    for (const name in form_input_elements) {
      toggle_view(form_input_elements[name], name === component_name)
    }
  }

  // -------------------------------
  // Protocol: program
  // -------------------------------

  function program_protocol (msg) {
    const program_handlers = {
      load_actions: program_load_actions
    }
    const { type, data } = msg
    const handler = program_handlers[type] || fail
    handler(data, type, msg)
  }

  function program_load_actions (data, type, msg) {
    variables = data
    _.up && _.up(type, data, msg.head ? { cause: msg.head } : {} )
  }

  // -------------------------------
  // Protocol: steps wizard
  // -------------------------------

  function steps_wizard_protocol (msg) {
    const steps_handlers = {
      step_clicked: steps_wizard_step_clicked
    }

    const { type } = msg
    const handler = steps_handlers[type]
    if (handler) handler(msg)
    else _.up && _.up(type, msg.data, msg.head ? { cause: msg.head } : {} )
  }

  function steps_wizard_step_clicked (msg) {
    const { data } = msg
    const refs = msg.head ? { cause: msg.head } : {} 
    _.up && _.up('step_clicked', data, refs)

    if (should_execute_step(data)) {
      _.up && _.up('execute_step', {
        action: selected_action,
        step: data,
        commands: data.commands
      }, refs)
    }

    function should_execute_step (step_data) {
      return step_data && Array.isArray(step_data.commands) && step_data.commands.length > 0
    }
  }

  // -------------------------------
  // Protocol: form input
  // -------------------------------

  function form_input_protocol (component_name) {
    return on

    function on (msg) {
      const form_input_handlers = {
        action_submitted: form_action_submitted,
        action_incomplete: form_action_incomplete,
        action_complete: form__action_complete
      }
      const { type, data } = msg
      const handler = form_input_handlers[type] || fail
      handler(data, type, msg)
    }
  }

  function form_action_submitted (data, type, msg) {
    console.log('exec.on_form_submitted', data, variables, selected_action)
    const step = selected_action.steps[data?.index]
    Object.assign(step, {
      is_completed: true,
      status: 'completed',
      data: data.value
    })

    const refs = msg.head ? { cause: msg.head } : {} 
    _.program('update_data', all_data, refs)
    _.steps_wizard('init_data', selected_action.steps, refs)

    if (selected_action.steps[selected_action.steps.length - 1]?.is_completed) {
      _.up && _.up('show_submit_btn', null, refs)
    }
  }

  function form_action_incomplete (data, type, msg) {
    console.log('exec.on_form_incomplete', data, variables, selected_action)
    const step = selected_action.steps[data?.index]

    if (!step.is_completed) return

    Object.assign(step, {
      is_completed: false,
      status: 'error',
      data: data.value !== undefined ? data.value : undefined
    })
    const refs = msg.head ? { cause: msg.head } : {} 
    _.program('update_data', all_data, refs)
    _.steps_wizard('init_data', selected_action.steps, refs)
    _.up && _.up('hide_submit_btn', null, refs)
  }

  function form__action_complete (data, type, msg) {
    if (!selected_action || !selected_action.steps) return

    const all_mandatory_complete = selected_action.steps.every(is_step_complete_or_optional)

    if (all_mandatory_complete) {
      _.up && _.up('action_auto_completed', { selected_action, trigger: 'form' }, msg.head ? { cause: msg.head } : {} )
    }

    function is_step_complete_or_optional (step) {
      return step.is_completed || step.type === 'optional'
    }
  }

  // -------------------------------
  // onmessage from parent
  // -------------------------------

  function onmessage (msg) {
    const { type } = msg
    if (type === 'docs_toggle') {
      _.steps_wizard(type, msg.data, msg.head ? { cause: msg.head } : {} )
      for (const name in component_modules) {
        _[name](type, msg.data, msg.head ? { cause: msg.head } : {} )
      }
    } else {
      parent_handler[type](msg)
    }
  }

  function update_steps_wizard_for_app (msg) {
    const { data } = msg
    all_data = data
  }

  function load_actions (msg) {
    variables = msg.data
    _.program(msg.type, msg.data, msg.head ? { cause: msg.head } : {} )
  }

  function action_submitted (msg) {
    const { data } = msg
    data.result = JSON.stringify(selected_action.steps.map(step => step.data), null, 2)

    reset_selected_action_steps()
    _.program('display_result', data, msg.head ? { cause: msg.head } : {} )
  }

  function reset_selected_action_steps () {
    if (!selected_action?.steps) return

    selected_action.steps.forEach(step => {
      step.data = ''
      step.is_completed = false
    })
  }

  function render_form (msg) {
    const { data } = msg
    render_form_component(data.component)
    const send = _[data.component]
    if (send) {
      send('step_data', data, msg.head ? { cause: msg.head } : {} )
    }
  }

  function parent_selected_action (msg) { selected_action = msg.data }

  function update_data (msg) {
    variables = msg.data
    const { data: msg_data, type } = msg
    _.program(type, msg_data, msg.head ? { cause: msg.head } : {} )
  }

  function activate_steps_wizard (msg) {
    if (!all_data) return
    const steps_data = all_data.find(matches_selected_action)
    selected_action = steps_data
    if (!steps_data) return
    steps_toggle_view('block')
    const data = steps_data.steps
    _.steps_wizard('init_data', data, msg.head ? { cause: msg.head } : {} )

    function matches_selected_action (action) { return action.name === msg.data }
  }

  function form_data (msg) {
    // forward init_data to steps_wizard with current action steps
    _.steps_wizard('init_data', msg.data, msg.head ? { cause: msg.head } : {} )
  }

  function clean_up (msg) {
    steps_toggle_view('none')
    for (const el of Object.values(form_input_elements)) {
      toggle_view(el, false)
    }
  }
}

function fallback_module () {
  return {
    api: fallback_instance,
    _: {
      program: { $: '' },
      steps_wizard: { $: '' },
      net_helper: { $: '' }
    }
  }
  function fallback_instance () {
    return {
      _: {
        program: {
          0: '',
          mapping: {
            style: 'style',
            variables: 'variables',
            docs: 'docs'
          }
        },
        steps_wizard: {
          0: '',
          mapping: {
            style: 'style',
            variables: 'variables',
            docs: 'docs'
          }
        },
        'program>form_input': {
          0: '',
          mapping: {
            style: 'style',
            data: 'data',
            docs: 'docs'
          }
        },
        'program>input_test': {
          0: '',
          mapping: {
            style: 'style',
            data: 'data',
            docs: 'docs'
          }
        },
        'program>form_tile_split_choice': {
          0: '',
          mapping: {
            style: 'style',
            data: 'data',
            docs: 'docs'
          },
          DOCS: {
            0: ''
          }
        },
        net_helper: {
          0: ''
        }
      },
      drive: {
        'style/': {
          'action_executor.css': {
            raw: `
              .container {
                display: flex;
                flex-direction: column;
                width: 100%;
              }
              .program-bar {
                display: flex;
              }
              .form-inputs {
                display: flex;
              }
              .steps-wizard-bar {
                display: flex;
              }
              .hide {
                display: none;
              }
            `
          }
        },
        'variables/': {},
        'data/': {},
        'docs/': {}
      }
    }
  }
}
