// action_executor/action_executor.js
const STATE = require('STATE')
const statedb = STATE(__filename)
const { get } = statedb(fallback_module)
const net = require('net_helper')

const program = require('program')
const steps_wizard = require('steps_wizard')

const { form_input, input_test, form_tile_split_choice } = program

const component_modules = {
  form_input,
  input_test,
  form_tile_split_choice
  // Add more form input components here if needed
}

module.exports = action_executor

async function action_executor (opts, protocol) {
  const { id, sdb } = await get(opts.sid)
  const { drive } = sdb
  const ids = opts.ids
  if (!ids || !ids.up) {
    throw new Error(`Component ${__filename} requires ids.up to be provided`)
  }
  const by = id
  const to = ids.up

  const on = {
    style: inject
  }

  const el = document.createElement('div')
  const shadow = el.attachShadow({ mode: 'closed' })
  shadow.innerHTML = `
  <div class="container main">
    <program></program>
    <form-input></form-input>
    <steps-wizard></steps-wizard>
  </div>`

  const sheet = new CSSStyleSheet()
  shadow.adoptedStyleSheets = [sheet]
  const program_placeholder = shadow.querySelector('program')
  const form_input_placeholder = shadow.querySelector('form-input')
  const steps_wizard_placeholder = shadow.querySelector('steps-wizard')
  const { io, _ } = net(id)

  const subs = await sdb.watch(onbatch)

  _.send_form_input = {}
  let all_data = null
  let variables = []
  let selected_action = null
  let mid = 0

  io.on.up = onmessage
  io.on.program = program_protocol
  io.on.steps_wizard = steps_wizard_protocol

  const program_sid = subs[0].sid
  const steps_wizard_sid = subs[1].sid
  const form_input_sids = {}

  // dynamic form input component SIDs
  for (const [index, [component_name]] of Object.entries(component_modules).entries()) {
    const final_index = index + 2
    const sub_entry = subs[final_index] || { sid: opts.sid }
    form_input_sids[component_name] = sub_entry.sid
    io.on[component_name] = form_input_protocol(component_name)
  }
  if (protocol) io.accept(protocol)

  const program_el = await program({ ...subs[0], ids: { up: id } }, io.invite('program', { up: id }))
  _.send_program = _.program
  program_el.classList.add('program-bar', 'hide')
  program_placeholder.replaceWith(program_el)

  const steps_wizard_el = await steps_wizard({ ...subs[1], ids: { up: id } }, io.invite('steps_wizard', { up: id }))
  _.send_steps_wizard = _.steps_wizard
  steps_wizard_el.classList.add('steps-wizard-bar', 'hide')
  steps_wizard_placeholder.replaceWith(steps_wizard_el)

  const form_input_elements = {}

  for (const [index, [component_name, component_fn]] of Object.entries(component_modules).entries()) {
    const final_index = index + 2
    const sub_entry = subs[final_index] || { sid: opts.sid }
    const element = await component_fn({ ...sub_entry, ids: { up: id } }, io.invite(component_name, { up: id }))
    _.send_form_input[component_name] = _[component_name]
    element.classList.add('form-inputs', 'hide')
    form_input_elements[component_name] = element
    form_input_placeholder.parentNode.insertBefore(element, form_input_placeholder)
  }

  form_input_placeholder.remove()

  const parent_handler = {
    update_steps_wizard_for_app,
    load_actions,
    action_submitted,
    update_data,
    activate_steps_wizard,
    form_data,
    render_form,
    selected_action: parent_selected_action,
    clean_up
  }

  return el

  async function onbatch (batch) {
    for (const { type, paths } of batch) {
      const data = await Promise.all(paths.map(load_path_raw))
      const func = on[type] || fail
      func(data, type)
    }

    function load_path_raw (path) { return drive.get(path).then(read_drive_file_raw) }
    function read_drive_file_raw (file) { return file.raw }
  }
  function fail (data, type) { console.warn('Unknown message type:', type, data) }
  function inject (data) { sheet.replaceSync(data[0]) }

  function send_up (msg) {
    if (protocol) _.up(msg)
  }

  // --- Toggle Views ---
  function toggle_view (el, show) { el.classList.toggle('hide', !show) }
  function steps_toggle_view (display) { toggle_view(steps_wizard_el, display === 'block') }

  function render_form_component (component_name) {
    for (const name in form_input_elements) {
      toggle_view(form_input_elements[name], name === component_name)
    }
  }

  // -------------------------------
  // Protocol: program
  // -------------------------------

  function program_protocol (msg) {
    const program_handlers = {
      load_actions: program_load_actions
    }
    const { type, data } = msg
    const handler = program_handlers[type] || fail
    handler(data, type, msg)
  }

  function program_load_actions (data, type, msg) {
    variables = data
    const head = [by, to, mid++]
    const refs = msg.head ? { cause: msg.head } : {}
    send_up({ head, refs, type, data })
  }

  // -------------------------------
  // Protocol: steps wizard
  // -------------------------------

  function steps_wizard_protocol (msg) {
    const steps_handlers = {
      step_clicked: steps_wizard_step_clicked
    }

    const { type } = msg
    const handler = steps_handlers[type]
    if (handler) handler(msg)
    else send_up(msg)
  }

  function steps_wizard_step_clicked (msg) {
    const { data } = msg
    const head = [by, to, mid++]
    const refs = msg.head ? { cause: msg.head } : {}
    send_up({ head, refs, type: 'step_clicked', data })

    if (should_execute_step(data)) {
      const execute_head = [by, to, mid++]
      send_up({
        head: execute_head,
        refs,
        type: 'execute_step',
        data: {
          action: selected_action,
          step: data,
          commands: data.commands
        }
      })
    }

    function should_execute_step (step_data) {
      return step_data && Array.isArray(step_data.commands) && step_data.commands.length > 0
    }
  }

  // -------------------------------
  // Protocol: form input
  // -------------------------------

  function form_input_protocol (component_name) {
    return on

    function on (msg) {
      const form_input_handlers = {
        action_submitted: form_action_submitted,
        action_incomplete: form_action_incomplete,
        action_complete: form__action_complete
      }
      const { type, data } = msg
      const handler = form_input_handlers[type] || fail
      handler(data, type, msg)
    }
  }

  function form_action_submitted (data, type, msg) {
    console.log('exec.on_form_submitted', data, variables, selected_action)
    const step = selected_action.steps[data?.index]
    Object.assign(step, {
      is_completed: true,
      status: 'completed',
      data: data.value
    })

    const refs = msg.head ? { cause: msg.head } : {}
    const head_to_program = [by, program_sid, mid++]
    _.send_program?.({ head: head_to_program, refs, type: 'update_data', data: all_data })

    const head_to_steps = [by, steps_wizard_sid, mid++]
    _.send_steps_wizard?.({ head: head_to_steps, type: 'init_data', data: selected_action.steps })

    if (selected_action.steps[selected_action.steps.length - 1]?.is_completed) {
      const head = [by, to, mid++]
      send_up({ head, refs, type: 'show_submit_btn' })
    }
  }

  function form_action_incomplete (data, type, msg) {
    console.log('exec.on_form_incomplete', data, variables, selected_action)
    const step = selected_action.steps[data?.index]

    if (!step.is_completed) return

    Object.assign(step, {
      is_completed: false,
      status: 'error',
      data: data.value !== undefined ? data.value : undefined
    })
    const refs = msg.head ? { cause: msg.head } : {}
    const head_to_program = [by, program_sid, mid++]
    _.send_program?.({ head: head_to_program, refs, type: 'update_data', data: all_data })

    const head_to_steps = [by, steps_wizard_sid, mid++]
    _.send_steps_wizard?.({ head: head_to_steps, type: 'init_data', data: selected_action.steps })

    const head = [by, to, mid++]
    send_up({ head, refs, type: 'hide_submit_btn' })
  }

  function form__action_complete (data, type, msg) {
    if (!selected_action || !selected_action.steps) return

    const all_mandatory_complete = selected_action.steps.every(is_step_complete_or_optional)

    if (all_mandatory_complete) {
      const refs = msg.head ? { cause: msg.head } : {}
      const head = [by, to, mid++]
      send_up({ head, refs, type: 'action_auto_completed', data: { selected_action, trigger: 'form' } })
    }

    function is_step_complete_or_optional (step) {
      return step.is_completed || step.type === 'optional'
    }
  }

  // -------------------------------
  // onmessage from parent
  // -------------------------------

  function onmessage (msg) {
    const { type } = msg
    if (type === 'docs_toggle') {
      _.send_steps_wizard(msg)
      for (const name in _.send_form_input) {
        _.send_form_input[name](msg)
      }
    } else {
      parent_handler[type](msg)
    }
  }

  function update_steps_wizard_for_app (msg) {
    const { data } = msg
    all_data = data
  }

  function load_actions (msg) {
    variables = msg.data
    const { data, type } = msg
    const head = [by, program_sid, mid++]
    const refs = msg.head ? { cause: msg.head } : {}
    _.send_program({ head, refs, type, data })
  }

  function action_submitted (msg) {
    const { data } = msg
    data.result = JSON.stringify(selected_action.steps.map(step => step.data), null, 2)

    reset_selected_action_steps()
    const head = [by, program_sid, mid++]
    const refs = msg.head ? { cause: msg.head } : {}
    _.send_program({ head, refs, type: 'display_result', data })
  }

  function reset_selected_action_steps () {
    if (!selected_action?.steps) return

    selected_action.steps.forEach(step => {
      step.data = ''
      step.is_completed = false
    })
  }

  function render_form (msg) {
    const { data } = msg
    render_form_component(data.component)
    const send = _.send_form_input[data.component]
    if (send) {
      const head = [by, form_input_sids[data.component], mid++]
      const refs = msg.head ? { cause: msg.head } : {}
      send({ head, refs, type: 'step_data', data })
    }
  }

  function parent_selected_action (msg) { selected_action = msg.data }

  function update_data (msg) {
    const { data, type } = msg
    variables = data
    const head = [by, program_sid, mid++]
    const refs = msg.head ? { cause: msg.head } : {}
    _.send_program({ head, refs, type, data })
  }

  function activate_steps_wizard (msg) {
    if (!all_data) return
    const steps_data = all_data.find(matches_selected_action)
    selected_action = steps_data
    if (!steps_data) return
    steps_toggle_view('block')
    const head_to_steps = [by, steps_wizard_sid, mid++]
    const data = steps_data.steps
    _.send_steps_wizard({
      head: head_to_steps,
      refs: {},
      type: 'init_data',
      data
    })

    function matches_selected_action (action) { return action.name === msg.data }
  }

  function form_data (msg) {
    // forward init_data to steps_wizard with current action steps
    const { data } = msg
    const head_to_steps = [by, steps_wizard_sid, mid++]
    _.send_steps_wizard({ head: head_to_steps, type: 'init_data', data })
  }

  function clean_up (msg) {
    steps_toggle_view('none')
    for (const el of Object.values(form_input_elements)) {
      toggle_view(el, false)
    }
  }
}

function fallback_module () {
  return {
    api: fallback_instance,
    _: {
      program: { $: '' },
      steps_wizard: { $: '' },
      net_helper: { $: '' }
    }
  }
  function fallback_instance () {
    return {
      _: {
        program: {
          0: '',
          mapping: {
            style: 'style',
            variables: 'variables',
            docs: 'docs'
          }
        },
        steps_wizard: {
          0: '',
          mapping: {
            style: 'style',
            variables: 'variables',
            docs: 'docs'
          }
        },
        'program>form_input': {
          0: '',
          mapping: {
            style: 'style',
            data: 'data',
            docs: 'docs'
          }
        },
        'program>input_test': {
          0: '',
          mapping: {
            style: 'style',
            data: 'data',
            docs: 'docs'
          }
        },
        'program>form_tile_split_choice': {
          0: '',
          mapping: {
            style: 'style',
            data: 'data',
            docs: 'docs'
          },
          DOCS: {
            0: ''
          }
        },
        net_helper: {
          0: ''
        }
      },
      drive: {
        'style/': {
          'action_executor.css': {
            raw: `
              .container {
                display: flex;
                flex-direction: column;
                width: 100%;
              }
              .program-bar {
                display: flex;
              }
              .form-inputs {
                display: flex;
              }
              .steps-wizard-bar {
                display: flex;
              }
              .hide {
                display: none;
              }
            `
          }
        },
        'variables/': {},
        'data/': {},
        'docs/': {}
      }
    }
  }
}
