// program_container/program_container.js
const STATE = require('STATE')
const statedb = STATE(__filename)
const { get } = statedb(fallback_module)
const DOCS = require('DOCS')
const docs = DOCS(__filename)()

const console_history = require('console_history')
const actions = require('actions')
const tabbed_editor = require('tabbed_editor')
const graph_explorer_wrapper = require('graph_explorer_wrapper')
const docs_window = require('docs_window')

module.exports = program_container

async function program_container (opts, protocol) {
  const { id, sdb } = await get(opts.sid)
  const { drive } = sdb
  const ids = opts.ids
  if (!ids || !ids.up) {
    throw new Error(`Component ${__filename} requires ids.up to be provided`)
  }
  const by = id
  const to = ids.up
  let mid = 0

  const on = {
    style: inject
  }

  const el = document.createElement('div')
  const shadow = el.attachShadow({ mode: 'closed' })
  shadow.innerHTML = `
  <div class="program-container main">
    <docs-window-placeholder></docs-window-placeholder>
    <graph-explorer-placeholder></graph-explorer-placeholder>
    <actions-placeholder></actions-placeholder>
    <tabbed-editor-placeholder></tabbed-editor-placeholder>
    <console-history-placeholder></console-history-placeholder>
  </div>
  <style>
  </style>`
  const style = shadow.querySelector('style')
  const graph_explorer_placeholder = shadow.querySelector('graph-explorer-placeholder')
  const actions_placeholder = shadow.querySelector('actions-placeholder')
  const tabbed_editor_placeholder = shadow.querySelector('tabbed-editor-placeholder')
  const console_placeholder = shadow.querySelector('console-history-placeholder')
  const docs_window_placeholder = shadow.querySelector('docs-window-placeholder')

  let console_history_el = null
  let docs_window_el = null
  let actions_el = null
  let tabbed_editor_el = null
  let graph_explorer_el = null

  const subs = await sdb.watch(onbatch)
  let send = null
  let _ = null
  if (protocol) {
    send = protocol(onmessage)
    _ = { up: send, actions: null, send_console_history: null, send_tabbed_editor: null, send_graph_explorer: null, send_docs_window: null }
  }

  docs_window_el = protocol ? await docs_window({ ...subs[4], ids: { up: id } }, docs_window_protocol) : await docs_window({ ...subs[4], ids: { up: id } })
  docs_window_el.classList.add('docs-window')
  docs_window_el.classList.add('hide')
  docs_window_placeholder.replaceWith(docs_window_el)

  graph_explorer_el = protocol ? await graph_explorer_wrapper({ ...subs[3], ids: { up: id } }, graph_explorer_protocol) : await graph_explorer_wrapper({ ...subs[3], ids: { up: id } })
  graph_explorer_el.classList.add('graph-explorer')
  graph_explorer_placeholder.replaceWith(graph_explorer_el)

  actions_el = protocol ? await actions({ ...subs[1], ids: { up: id } }, actions_protocol) : await actions({ ...subs[1], ids: { up: id } })
  actions_el.classList.add('actions')
  actions_placeholder.replaceWith(actions_el)

  tabbed_editor_el = protocol ? await tabbed_editor({ ...subs[2], ids: { up: id } }, tabbed_editor_protocol) : await tabbed_editor({ ...subs[2], ids: { up: id } })
  tabbed_editor_el.classList.add('tabbed-editor')
  tabbed_editor_placeholder.replaceWith(tabbed_editor_el)

  console_history_el = protocol ? await console_history({ ...subs[0], ids: { up: id } }, console_history_protocol) : await console_history({ ...subs[0], ids: { up: id } })
  console_history_el.classList.add('console-history')
  console_placeholder.replaceWith(console_history_el)
  let console_view = false
  let actions_view = false
  let graph_explorer_view = false

  if (protocol) {
    console_history_el.classList.add('hide')
    actions_el.classList.add('hide')
    tabbed_editor_el.classList.add('show')
    graph_explorer_el.classList.add('hide')

    // Send message to root to set doc display handler
    _.up({
      head: [by, to, mid++],
      refs: {},
      type: 'set_doc_display_handler',
      data: {
        callback: on_doc_display
      }
    })
  }

  return el

  function console_history_toggle_view () {
    if (console_view) {
      console_history_el.classList.remove('show')
      console_history_el.classList.add('hide')
    } else {
      console_history_el.classList.remove('hide')
      console_history_el.classList.add('show')
    }
    console_view = !console_view
  }

  function actions_toggle_view () {
    if (actions_view) {
      actions_el.classList.remove('show')
      actions_el.classList.add('hide')
    } else {
      actions_el.classList.remove('hide')
      actions_el.classList.add('show')
    }
    actions_view = !actions_view
  }

  function graph_explorer_toggle_view () {
    if (graph_explorer_view) {
      graph_explorer_el.classList.remove('show')
      graph_explorer_el.classList.add('hide')
    } else {
      graph_explorer_el.classList.remove('hide')
      graph_explorer_el.classList.add('show')
    }
    graph_explorer_view = !graph_explorer_view
  }

  function tabbed_editor_toggle_view (show = true) {
    if (show) {
      tabbed_editor_el.classList.remove('hide')
      tabbed_editor_el.classList.add('show')
      actions_el.classList.remove('show')
      actions_el.classList.add('hide')
      console_history_el.classList.remove('show')
      console_history_el.classList.add('hide')
      graph_explorer_el.classList.remove('show')
      graph_explorer_el.classList.add('hide')
      actions_view = false
      console_view = false
      graph_explorer_view = false
    } else {
      tabbed_editor_el.classList.remove('show')
      tabbed_editor_el.classList.add('hide')
    }
  }

  async function onbatch (batch) {
    for (const { type, paths } of batch) {
      const data = await Promise.all(paths.map(load_path_raw))
      const func = on[type] || fail
      func(data, type)
    }

    function load_path_raw (path) { return drive.get(path).then(read_drive_file_raw) }
    function read_drive_file_raw (file) { return file.raw }
  }
  function fail (data, type) { console.warn('invalid message', { cause: { data, type } }) }
  function inject (data) {
    style.replaceChildren(create_style_element())

    function create_style_element () {
      const style = document.createElement('style')
      style.textContent = data[0]
      return style
    }
  }

  function on_doc_display (display_data) {
    const { content, sid } = display_data
    docs_window_el.classList.remove('hide')
    if (_.send_docs_window) {
      _.send_docs_window({ type: 'display_doc', data: { content, sid } })
    }
  }

  // ---------
  // PROTOCOLS
  // ---------

  function console_history_protocol (send) {
    _.send_console_history = send
    return on
    function on (msg) { _.up(msg) }
  }

  function actions_protocol (send) {
    _.send_actions = send

    const action_handlers = {
      selected_action: actions_selected_action,
      ui_focus_docs: actions_ui_focus_docs,
      ui_focus: actions__forward_up
    }

    return on
    function on (msg) {
      const handler = action_handlers[msg.type] || actions__forward_up
      handler(msg)
    }

    function actions__forward_up (msg) { _.up && _.up(msg) }
  }

  function actions_selected_action (msg) {
    const { data } = msg
    const head = [by, to, mid++]
    const refs = msg.head ? { cause: msg.head } : {}
    _.up({
      head,
      refs,
      type: 'update_quick_actions_input',
      data
    })
  }

  function actions_ui_focus_docs (msg) { _.up(msg) }

  function tabbed_editor_protocol (send) {
    _.send_tabbed_editor = send
    return on
    function on (msg) { _.up(msg) }
  }

  function graph_explorer_protocol (send) {
    _.send_graph_explorer = send
    return on
    function on (msg) { _.up(msg) }
  }

  function docs_window_protocol (send) {
    _.send_docs_window = send
    const action_handlers = {
      close_docs: docs_window__close_docs
    }
    return on
    function on (msg) {
      const handler = action_handlers[msg.type] || docs_window__noop
      handler(msg)
      _.up && _.up(msg)
    }

    function docs_window__close_docs () { docs_window_el.classList.add('hide') }

    function docs_window__noop () {}
  }

  function onmessage (msg) {
    const action_handlers = {
      console_history_toggle: onmessage__console_history_toggle,
      graph_explorer_toggle: onmessage__graph_explorer_toggle,
      display_actions: onmessage__display_actions,
      filter_actions: onmessage__filter_actions,
      tab_name_clicked: onmessage__tab_name_clicked,
      tab_close_clicked: onmessage__tab_close_clicked,
      switch_tab: onmessage__switch_tab,
      entry_toggled: onmessage__entry_toggled,
      display_doc: onmessage__display_doc,
      load_actions: onmessage__send_actions,
      update_actions_for_app: onmessage__send_actions
    }
    const handler = action_handlers[msg.type] || onmessage__noop
    handler(msg)

    function onmessage__console_history_toggle () { console_history_toggle_view() }
    function onmessage__graph_explorer_toggle () { graph_explorer_toggle_view() }
    function onmessage__display_actions (msg) { actions_toggle_view(msg.data) }
    function onmessage__filter_actions (msg) { _.send_actions && _.send_actions(msg) }
    function onmessage__tab_close_clicked (msg) { _.send_tabbed_editor && _.send_tabbed_editor({ ...msg, type: 'close_tab' }) }
    function onmessage__entry_toggled (msg) { _.send_graph_explorer && _.send_graph_explorer(msg) }
    function onmessage__send_actions (msg) { _.send_actions && _.send_actions(msg) }
    function onmessage__noop () {}
    function onmessage__tab_name_clicked (msg) {
      tabbed_editor_toggle_view(true)
      _.send_tabbed_editor && _.send_tabbed_editor({ ...msg, type: 'toggle_tab' })
    }
    function onmessage__switch_tab (msg) {
      tabbed_editor_toggle_view(true)
      _.send_tabbed_editor && _.send_tabbed_editor(msg)
    }
    function onmessage__display_doc (msg) {
      docs_window_el.classList.remove('hide')
      _.send_docs_window && _.send_docs_window(msg)
    }
  }
}

function fallback_module () {
  return {
    api: fallback_instance,
    _: {
      console_history: {
        $: ''
      },
      actions: {
        $: ''
      },
      tabbed_editor: {
        $: ''
      },
      graph_explorer_wrapper: {
        $: ''
      },
      docs_window: {
        $: ''
      },
      DOCS: {
        $: ''
      }
    }
  }

  function fallback_instance () {
    return {
      _: {
        console_history: {
          0: '',
          mapping: {
            style: 'style',
            commands: 'commands',
            icons: 'icons',
            scroll: 'scroll',
            docs: 'docs',
            actions: 'actions'
          }
        },
        actions: {
          0: '',
          mapping: {
            style: 'style',
            actions: 'actions',
            icons: 'icons',
            hardcons: 'hardcons',
            docs: 'docs'
          }
        },
        tabbed_editor: {
          0: '',
          mapping: {
            style: 'style',
            files: 'files',
            highlight: 'highlight',
            active_tab: 'active_tab',
            docs: 'docs'
          }
        },
        graph_explorer_wrapper: {
          0: '',
          mapping: {
            theme: 'style',
            entries: 'entries',
            runtime: 'runtime',
            mode: 'mode',
            flags: 'flags',
            keybinds: 'keybinds',
            undo: 'undo',
            docs: 'docs'
          }
        },
        docs_window: {
          0: '',
          mapping: {
            style: 'docs_style'
          }
        },
        DOCS: {
          0: ''
        }
      },
      drive: {
        'style/': {
          'theme.css': {
            raw: `
              .program-container {
                display: grid;
                grid-template-rows: 1fr auto auto;
                min-height: 200px;
                background: linear-gradient(135deg, #0d1117 0%, #161b22 100%);
                position: relative;
                gap: 8px;
                padding: 8px;
                height: stretch;
                padding-bottom: 0px;
              }
              .console-history {
                grid-row: 3;
                position: relative;
                width: 100%;
                background-color: #161b22;
                border: 1px solid #21262d;
                border-radius: 6px;
              }
              .actions {
                grid-row: 2;
                position: relative;
                width: 100%;
                background-color: #161b22;
                border: 1px solid #21262d;
                border-radius: 6px;
              }
              .tabbed-editor {
                grid-row: 1;
                height: min-content;
                position: relative;
                width: 100%;
                background-color: #0d1117;
                border: 1px solid #21262d;
                border-radius: 6px;
                overflow: hidden;
              }
              .show {
                display: block;
              }
              .hide {
                display: none;
              }
            `
          }
        },
        'entries/': {},
        'flags/': {},
        'keybinds/': {},
        'commands/': {},
        'icons/': {},
        'scroll/': {},
        'actions/': {},
        'hardcons/': {},
        'files/': {},
        'highlight/': {},
        'active_tab/': {},
        'runtime/': {},
        'mode/': {},
        'undo/': {},
        'docs_style/': {}
      }
    }
  }
}
