// manager/manager.js
const STATE = require('STATE')
const statedb = STATE(__filename)
const { get } = statedb(fallback_module)

const program = require('program')
const action_bar = require('action_bar')

const { form_input, input_test } = program

const component_modules = {
  form_input,
  input_test
  // Add more form input components here if needed
}

module.exports = manager

async function manager (opts, protocol) {
  const { id, sdb } = await get(opts.sid)
  const { drive } = sdb
  const ids = opts.ids
  if (!ids || !ids.up) {
    throw new Error(`Component ${__filename} requires ids.up to be provided`)
  }
  const by = id
  // const to = ids.up

  const on = {
    style: inject
  }

  let variables = []
  let selected_action = null
  let mid = 0

  let _ = null
  if (protocol) {
    const send = protocol(msg => onmessage(msg))
    _ = { up: send, send_actions_bar: null, send_form_input: {}, send_program: null }
  } else {
    _ = { send_actions_bar: null, send_form_input: {}, send_program: null }
  }

  const el = document.createElement('div')
  const shadow = el.attachShadow({ mode: 'closed' })
  shadow.innerHTML = `
    <div class="main">
      <form-input></form-input>
      <action-bar></action-bar>
      <program></program>
    </div>
    <style></style>
  `

  const style = shadow.querySelector('style')
  const form_input_placeholder = shadow.querySelector('form-input')
  const program_placeholder = shadow.querySelector('program')
  const action_bar_placeholder = shadow.querySelector('action-bar')

  const subs = await sdb.watch(onbatch)

  const action_bar_sid = subs[0].sid
  const program_sid = subs[1].sid
  const form_input_sids = {}

  // dynamic form input component SIDs
  for (const [index, [component_name]] of Object.entries(component_modules).entries()) {
    const final_index = index + 2
    form_input_sids[component_name] = subs[final_index].sid
  }

  const action_bar_el = await action_bar({ ...subs[0], ids: { up: id } }, actions_bar_protocol)
  action_bar_placeholder.replaceWith(action_bar_el)

  const program_el = await program({ ...subs[1], ids: { up: id } }, program_protocol)
  program_el.classList.add('hide')
  program_placeholder.replaceWith(program_el)

  const form_input_elements = {}

  console.log('subs', subs)

  for (const [index, [component_name, component_fn]] of Object.entries(component_modules).entries()) {
    const final_index = index + 2

    console.log('final_index', final_index, component_name, subs[final_index])

    const el = await component_fn({ ...subs[final_index], ids: { up: id } }, form_input_protocol(component_name))
    el.classList.add('hide')
    form_input_elements[component_name] = el
    form_input_placeholder.parentNode.insertBefore(el, form_input_placeholder)
  }

  form_input_placeholder.remove()

  return el

  function onmessage (msg) {
    const { type } = msg
    switch (type) {
    case 'update_actions_for_app':
      const head_to_action_bar = [by, action_bar_sid, mid++]
      const refs = msg.head ? { cause: msg.head } : {}
      _.send_actions_bar?.({ head: head_to_action_bar, refs, type, data: msg.data })
      break
    default: // @TODO Handle message types
    }
  }

  // --- Internal Functions ---
  async function onbatch (batch) {
    for (const { type, paths } of batch) {
      const data = await Promise.all(paths.map(path => drive.get(path).then(file => file.raw)))
      const func = on[type] || fail
      func(data, type)
    }
  }

  function fail (data, type) {
    throw new Error('invalid message', { cause: { data, type } })
  }

  function inject (data) {
    style.replaceChildren((() => {
      const style_el = document.createElement('style')
      style_el.textContent = data[0]
      return style_el
    })())
  }

  function toggle_view (el, show) {
    el.classList.toggle('hide', !show)
  }

  function render_form_component (component_name) {
    for (const name in form_input_elements) {
      toggle_view(form_input_elements[name], name === component_name)
    }
  }

  // -------------------------------
  // Protocol: form input
  // -------------------------------

  function form_input_protocol (component_name) {
    return function (send) {
      _.send_form_input[component_name] = send

      const form_input_handlers = {
        action_submitted: form__action_submitted,
        action_incomplete: form__action_incomplete
      }

      return on
      function on (msg) {
        const { type, data } = msg
        const handler = form_input_handlers[type] || fail
        handler(data, type, msg)
      }
    }
  }

  function form__action_submitted (data, type, msg) {
    console.log('manager.on_form_submitted', data, variables, selected_action)
    const step = variables[selected_action][data?.index]
    Object.assign(step, {
      is_completed: true,
      status: 'completed',
      data: data?.value
    })
    const head_to_program = [by, program_sid, mid++]
    const refs = msg.head ? { cause: msg.head } : {}
    _.send_program?.({ head: head_to_program, refs, type: 'update_data', data: variables })

    const head_to_action_bar = [by, action_bar_sid, mid++]
    _?.send_actions_bar({ head: head_to_action_bar, refs, type: 'form_data', data: variables[selected_action] })

    if (variables[selected_action][variables[selected_action].length - 1]?.is_completed) {
      const head_to_action_bar_submit = [by, action_bar_sid, mid++]
      _.send_actions_bar({ head: head_to_action_bar_submit, refs, type: 'show_submit_btn' })
    }
  }

  function form__action_incomplete (data, type, msg) {
    console.log('manager.on_form_incomplete', data, variables, selected_action)
    const step = variables[selected_action][data?.index]

    if (!step.is_completed) return

    Object.assign(step, {
      is_completed: false,
      status: 'error',
      data: data?.value
    })
    const head_to_program = [by, program_sid, mid++]
    const refs = msg.head ? { cause: msg.head } : {}
    _.send_program?.({ head: head_to_program, refs, type: 'update_data', data: variables })

    const head_to_action_bar = [by, action_bar_sid, mid++]
    _?.send_actions_bar({ head: head_to_action_bar, refs, type: 'form_data', data: variables[selected_action] })

    const head_to_action_bar_hide = [by, action_bar_sid, mid++]
    _.send_actions_bar({ head: head_to_action_bar_hide, refs, type: 'hide_submit_btn' })
  }

  // -------------------------------
  // Protocol: program
  // -------------------------------

  function program_protocol (send) {
    _.send_program = send

    const program_handlers = {
      load_actions: program__load_actions
    }
    return function on (msg) {
      const { type, data } = msg
      const handler = program_handlers[type] || fail
      handler(data, type, msg)
    }
  }

  function program__load_actions (data, type, msg) {
    variables = data
    const head = [by, action_bar_sid, mid++]
    const refs = msg.head ? { cause: msg.head } : {}
    _.send_actions_bar?.({ head, refs, type, data })
  }

  // -------------------------------
  // Protocol: action bar
  // -------------------------------

  function actions_bar_protocol (send) {
    _.send_actions_bar = send

    const action_bar_handlers = {
      render_form: action_bar__render_form,
      clean_up: action_bar__clean_up,
      action_submitted: action_bar__action_submitted,
      selected_action: action_bar__selected_action
    }

    return function on (msg) {
      if (msg.type === "console_history_toggle" || msg.type === "ui_focus"){
        _.up(msg)
        return
      }

      const { type, data } = msg
      const handler = action_bar_handlers[type] || fail
      handler(data, type, msg)
    }
  }

  function action_bar__render_form (data, type, msg) {
    render_form_component(data.component)
    const send = _.send_form_input[data.component]
    if (send) {
      const head = [by, form_input_sids[data.component], mid++]
      const refs = msg.head ? { cause: msg.head } : {}
      send({ head, refs, type: 'step_data', data })
    }
  }

  function action_bar__action_submitted (data, type, msg) {
    const head = [by, program_sid, mid++]
    const refs = msg.head ? { cause: msg.head } : {}
    _.send_program({ head, refs, type: 'display_result', data })
  }

  function action_bar__selected_action (data, type, msg) {
    selected_action = data
  }

  function action_bar__clean_up (data, type, msg) {
    data && cleanup(data, msg)
  }

  function cleanup (selected_action, msg) {
    const cleaned = variables[selected_action].map(step => ({
      ...step,
      is_completed: false,
      data: ''
    }))
    variables[selected_action] = cleaned
    const head_to_program = [by, program_sid, mid++]
    const refs = msg?.head ? { cause: msg.head } : {}
    _.send_program?.({ head: head_to_program, refs, type: 'update_data', data: variables })

    for (const step of variables[selected_action]) {
      if (step.component && _.send_form_input[step.component]) {
        const head_to_input = [by, form_input_sids[step.component], mid++]
        _.send_form_input[step.component]({ head: head_to_input, refs, type: 'reset_data' })
      }
    }

    for (const el of Object.values(form_input_elements)) {
      console.log('toggle_view', el, false)
      toggle_view(el, false)
    }
  }
}

// --- Fallback Module ---
function fallback_module () {
  return {
    api: fallback_instance,
    _: {
      action_bar: { $: '' },
      program: { $: '' }
    }
  }

  function fallback_instance () {
    return {
      _: {
        action_bar: {
          0: '',
          mapping: {
            icons: 'icons',
            style: 'style'
          }
        },
        program: {
          0: '',
          mapping: {
            style: 'style',
            variables: 'variables'
          }
        },
        'program>form_input': {
          0: '',
          mapping: {
            style: 'style',
            data: 'data'
          }
        },
        'program>input_test': {
          0: '',
          mapping: {
            style: 'style',
            data: 'data'
          }
        }
      },
      drive: {
        'style/': {
          'manager.css': {
            raw: `
              .main {
                display: flex;
                flex-direction: column;
                width: 100%;
                height: 100%;
                background: #131315;
              }
              .hide {
                display: none;
              }
            `
          }
        }
      }
    }
  }
}
