// manager/manager.js
const STATE = require('STATE')
const statedb = STATE(__filename)
const { get } = statedb(fallback_module)
const DOCS = require('DOCS')

const action_bar = require('action_bar')

module.exports = manager

async function manager (opts, protocol) {
  const { id, sdb } = await get(opts.sid)
  const { drive } = sdb
  const ids = opts.ids
  if (!ids || !ids.up) {
    throw new Error(`Component ${__filename} requires ids.up to be provided`)
  }
  const by = id
  // const to = ids.up

  const on = {
    style: inject
  }

  let mid = 0
  const on_message = {
    docs_toggle: handle_docs_toggle,
    load_actions: handle_load_actions,
    step_clicked: handle_step_clicked,
    update_quick_actions_for_app: handle_update_quick_actions_for_app,
    update_quick_actions_input: handle_update_quick_actions_input,
    show_submit_btn: handle_submit_btn_toggle,
    hide_submit_btn: handle_submit_btn_toggle
  }

  let _ = null
  if (protocol) {
    const send = protocol(msg => onmessage(msg))
    _ = { up: send, send_actions_bar: null }
  } else {
    _ = { send_actions_bar: null }
  }

  const el = document.createElement('div')
  const shadow = el.attachShadow({ mode: 'closed' })
  shadow.innerHTML = `
    <action-bar></action-bar>
    <style></style>
  `

  const style = shadow.querySelector('style')
  const action_bar_placeholder = shadow.querySelector('action-bar')

  const subs = await sdb.watch(onbatch)

  const action_bar_sid = subs[0].sid

  const action_bar_el = await action_bar({ ...subs[0], ids: { up: id } }, actions_bar_protocol)
  action_bar_el.classList.add('main')
  action_bar_placeholder.replaceWith(action_bar_el)

  return el

  function onmessage (msg) {
    const handler = on_message[msg.type] || onmessage_fail
    handler(msg)
  }

  function handle_docs_toggle (msg) {
    _.send_actions_bar && _.send_actions_bar(msg)
  }

  function handle_load_actions (msg) {
    forward_to_action_bar(msg.type, msg.data, msg)
  }

  function handle_step_clicked (msg) {
    _.send_actions_bar && _.send_actions_bar(msg)
  }

  function handle_update_quick_actions_for_app (msg) {
    forward_to_action_bar(msg.type, msg.data, msg)
  }

  function handle_update_quick_actions_input (msg) {
    forward_to_action_bar(msg.type, msg.data, msg)
  }

  function handle_submit_btn_toggle (msg) {
    _.send_actions_bar && _.send_actions_bar(msg)
  }

  function onmessage_fail () {
    // @TODO Handle unsupported message types
  }

  function forward_to_action_bar (type, data, msg) {
    const head = [by, action_bar_sid, mid++]
    const refs = msg.head ? { cause: msg.head } : {}
    _.send_actions_bar && _.send_actions_bar({ head, refs, type, data })
  }

  // --- Internal Functions ---
  async function onbatch (batch) {
    for (const { type, paths } of batch) {
      const data = await Promise.all(paths.map(path => drive.get(path).then(file => file.raw)))
      const func = on[type] || fail
      func(data, type)
    }
  }

  function fail (data, type) {
    console.warn('invalid message', { cause: { data, type } })
  }

  function inject (data) {
    style.replaceChildren((() => {
      const style_el = document.createElement('style')
      style_el.textContent = data[0]
      return style_el
    })())
  }

  // -------------------------------
  // Protocol: action bar
  // -------------------------------

  function actions_bar_protocol (send) {
    _.send_actions_bar = send

    const action_handlers = {
      render_form: action_bar__render_form,
      clean_up: action_bar__clean_up,
      action_submitted: action_bar__action_submitted,
      selected_action: action_bar__selected_action,
      activate_steps_wizard: action_bar__activate_steps_wizard,
      console_history_toggle: action_bar__forward_up,
      ui_focus: action_bar__forward_up,
      display_actions: action_bar__forward_up,
      filter_actions: action_bar__forward_up
    }

    return function on (msg) {
      const { type, data } = msg
      const handler = action_handlers[type] || action_bar_fail
      handler(data, type, msg)
    }
  }

  function action_bar__forward_up (data, type, msg) {
    _.up && _.up(msg)
  }

  function action_bar_fail (data, type) {
    fail(data, type)
  }

  function action_bar__render_form (data, type, msg) {
    _.up && _.up(msg)
  }

  function action_bar__action_submitted (data, type, msg) {
    _.up && _.up(msg)
  }

  function action_bar__selected_action (data, type, msg) {
    _.up && _.up(msg)
  }

  function action_bar__activate_steps_wizard (data, type, msg) {
    _.up && _.up(msg)
  }

  function action_bar__clean_up (data, type, msg) {
    _.up && _.up(msg)
  }
}

// --- Fallback Module ---
function fallback_module () {
  return {
    api: fallback_instance,
    _: {
      action_bar: { $: '' },
      DOCS: { $: '' }
    }
  }

  function fallback_instance () {
    return {
      _: {
        action_bar: {
          0: '',
          mapping: {
            icons: 'icons',
            style: 'style',
            actions: 'actions',
            variables: 'variables',
            hardcons: 'hardcons',
            prefs: 'prefs',
            docs: 'docs'
          }
        }
      },
      drive: {
        'style/': {
          'manager.css': {
            raw: `
              .main {
                display: flex;
                flex-direction: column;
                justify-content: center;
                align-items: flex-start;
                width: 100%;
                height: 100%;
                background: #131315;
              }
              .hide {
                display: none;
              }
            `
          }
        },
        'variables/': {},
        'data/': {},
        'actions/': {},
        'hardcons/': {},
        'prefs/': {},
        'docs/': {
          'README.md': {
            $ref: 'README.md'
          }
        }
      }
    }
  }
}
