// tab_group/tab_group.js
const STATE = require('STATE')
const statedb = STATE(__filename)
const { get } = statedb(fallback_module)
const net = require('net_helper')

const tabs = require('tabs')
const program_container = require('program_container')
const action_bar = require('action_bar')
const action_executor = require('action_executor')

module.exports = tab_group

async function tab_group (opts, invite) {
  console.error('tab_group: initializing with opts', opts.sid)
  const { id, sdb } = await get(opts.sid)
  const { drive } = sdb

  const on = {
    style: inject
  }

  const { io, _ } = net(id)

  const el = document.createElement('div')
  const shadow = el.attachShadow({ mode: 'closed' })
  shadow.innerHTML = `
  <div class="tab-group">
    <div class="program-container-slot"></div>
    <div class="action-executor-slot"></div>
    <div class="action-bar-slot"></div>
    <div class="tabs-slot"></div>
  </div>
  `
  const sheet = new CSSStyleSheet()
  shadow.adoptedStyleSheets = [sheet]

  const program_container_slot = shadow.querySelector('.program-container-slot')
  const action_executor_slot = shadow.querySelector('.action-executor-slot')
  const action_bar_slot = shadow.querySelector('.action-bar-slot')
  const tabs_slot = shadow.querySelector('.tabs-slot')

  const subs = await sdb.watch(onbatch)
  console.error('tab_group: subs ready', subs)

  let tabs_el = null
  let program_container_el = null
  let action_bar_el = null
  let action_executor_el = null

  io.on = {
    up: protocol_from_root(),
    tabs: protocol_from_tabs(),
    program_container: protocol_from_program_container(),
    action_bar: protocol_from_action_bar(),
    action_executor: protocol_from_action_executor()
  }
  if (invite) {
    console.error('tab_group: accepting invite')
    io.accept(invite)
  }

  console.error('tab_group: creating program_container')
  program_container_el = await program_container({ ...subs[0], ids: { up: id } }, io.invite('program_container', { up: id }))
  program_container_el.classList.add('program-container')
  program_container_slot.replaceWith(program_container_el)
  console.error('tab_group: program_container created')

  console.error('tab_group: creating action_executor')
  action_executor_el = await action_executor({ ...subs[1], ids: { up: id } }, io.invite('action_executor', { up: id }))
  action_executor_el.classList.add('action-executor')
  action_executor_slot.replaceWith(action_executor_el)
  console.error('tab_group: action_executor created')

  console.error('tab_group: creating tabs')
  tabs_el = await tabs({ ...subs[2], ids: { up: id } }, io.invite('tabs', { up: id }))
  tabs_el.classList.add('tabs')
  tabs_slot.replaceWith(tabs_el)
  console.error('tab_group: tabs created')

  console.error('tab_group: creating action_bar')
  action_bar_el = await action_bar({ ...subs[3], ids: { up: id } }, io.invite('action_bar', { up: id }))
  action_bar_el.classList.add('action-bar')
  action_bar_slot.replaceWith(action_bar_el)
  console.error('tab_group: action_bar created')

  console.error('tab_group: initialization complete')
  return el

  // ---------------------------
  // BATCH HANDLER
  // ---------------------------

  async function onbatch (batch) {
    for (const { type, paths } of batch) {
      const data = await Promise.all(paths.map(load_path_raw))
      const func = on[type] || fail
      func({ data, type })
    }

    function load_path_raw (path) { return drive.get(path).then(read_drive_file_raw) }
    function read_drive_file_raw (file) { return file.raw }
  }

  function inject ({ data }) { sheet.replaceSync(data[0]) }

  function fail ({ data, type }) { console.error('tab_group: unhandled batch message', { type, data }) }

  // ---------------------------
  // MESSAGE FROM ROOT (tile_manager)
  // ---------------------------

  function protocol_from_root () {
    const on = {
      update_quick_actions_for_app,
      update_steps_wizard_for_app,
      update_actions_for_app: forward_to_program_container,
      load_actions: forward_to_program_container,
      create_default_tab
    }
    return function onmessage (msg) {
      console.error('tab_group: message from root', msg.type)
      ;(on[msg.type] || onfail)(msg)
    }
    function update_quick_actions_for_app (msg) {
      console.error('tab_group: forwarding to action_bar', msg.type)
      const refs = msg.head ? { cause: msg.head } : {}
      _.action_bar?.(msg.type, msg.data, refs)
    }
    function update_steps_wizard_for_app (msg) {
      console.error('tab_group: forwarding to action_executor', msg.type)
      const refs = msg.head ? { cause: msg.head } : {}
      _.action_executor?.(msg.type, msg.data, refs)
    }
    function forward_to_program_container (msg) {
      console.error('tab_group: forwarding to program_container', msg.type)
      const refs = msg.head ? { cause: msg.head } : {}
      _.program_container?.(msg.type, msg.data, refs)
    }
    function create_default_tab (msg) {
      console.error('tab_group: creating default tab', msg.data)
      const refs = msg.head ? { cause: msg.head } : {}
      _.tabs?.('add_default_tab', msg.data, refs)
    }
    function onfail (msg) { console.error('tab_group: unknown root message', msg) }
  }

  // ---------------------------
  // PROTOCOLS
  // ---------------------------

  function protocol_from_tabs () {
    return function onmessage (msg) {
      console.error('tab_group: tabs_protocol', msg.type, msg.data)
      const refs = msg.head ? { cause: msg.head } : {}
      _.up(msg.type, msg.data, refs)
    }
  }

  function protocol_from_program_container () {
    const forward_up = {
      ui_focus: forward,
      action_auto_completed: forward,
      action_complete: forward
    }
    return function onmessage (msg) {
      console.error('tab_group: program_container_protocol', msg.type)
      ;(forward_up[msg.type] || onfail)(msg)
    }
    function forward (msg) {
      const refs = msg.head ? { cause: msg.head } : {}
      _.up(msg.type, msg.data, refs)
    }
    function onfail (msg) { console.error('tab_group: unhandled program_container msg', msg) }
  }

  function protocol_from_action_bar () {
    const forward_to_executor = {
      action_submitted: to_executor,
      selected_action: to_executor,
      activate_steps_wizard: to_executor,
      render_form: to_executor,
      clean_up: to_executor
    }
    const forward_up = {
      ui_focus: to_up,
      display_actions: to_up,
      filter_actions: to_up,
      console_history_toggle: to_up
    }
    return function onmessage (msg) {
      console.error('tab_group: action_bar_protocol', msg.type)
      const handler = forward_to_executor[msg.type] || forward_up[msg.type] || onfail
      handler(msg)
    }
    function to_executor (msg) {
      const refs = msg.head ? { cause: msg.head } : {}
      _.action_executor?.(msg.type, msg.data, refs)
    }
    function to_up (msg) {
      const refs = msg.head ? { cause: msg.head } : {}
      _.up(msg.type, msg.data, refs)
    }
    function onfail (msg) { console.error('tab_group: unhandled action_bar msg', msg) }
  }

  function protocol_from_action_executor () {
    const forward_to_action_bar = {
      selected_action: to_action_bar,
      show_submit_btn: to_action_bar,
      hide_submit_btn: to_action_bar,
      update_quick_actions_input: to_action_bar,
      step_clicked: to_action_bar,
      load_actions: to_action_bar
    }
    const forward_up = {
      action_auto_completed: to_up_and_bar,
      action_complete: to_up
    }
    return function onmessage (msg) {
      console.error('tab_group: action_executor_protocol', msg.type)
      const handler = forward_to_action_bar[msg.type] || forward_up[msg.type] || onfail
      handler(msg)
    }
    function to_action_bar (msg) {
      const refs = msg.head ? { cause: msg.head } : {}
      _.action_bar?.(msg.type, msg.data, refs)
    }
    function to_up (msg) {
      const refs = msg.head ? { cause: msg.head } : {}
      _.up(msg.type, msg.data, refs)
    }
    function to_up_and_bar (msg) {
      const refs = msg.head ? { cause: msg.head } : {}
      _.action_bar?.('action_submitted', msg.data, refs)
      _.up(msg.type, msg.data, refs)
    }
    function onfail (msg) { console.error('tab_group: unhandled action_executor msg', msg) }
  }
}

function fallback_module () {
  return {
    api: fallback_instance,
    _: {
      tabs: { $: '' },
      program_container: { $: '' },
      action_bar: { $: '' },
      action_executor: { $: '' },
      net_helper: { $: '' }
    }
  }

  function fallback_instance () {
    return {
      _: {
        program_container: {
          0: '',
          mapping: {
            style: 'style',
            icons: 'icons',
            commands: 'commands',
            scroll: 'scroll',
            actions: 'actions',
            hardcons: 'hardcons',
            files: 'files',
            highlight: 'highlight',
            active_tab: 'active_tab',
            entries: 'entries',
            runtime: 'runtime',
            mode: 'mode',
            flags: 'flags',
            keybinds: 'keybinds',
            undo: 'undo',
            docs_style: 'docs_style',
            docs: 'docs'
          }
        },
        action_executor: {
          0: '',
          mapping: {
            style: 'style',
            icons: 'icons',
            actions: 'actions',
            docs: 'docs',
            data: 'data',
            hardcons: 'hardcons',
            variables: 'variables'
          }
        },
        tabs: {
          0: '',
          mapping: {
            style: 'style',
            icons: 'icons',
            scroll: 'scroll',
            actions: 'actions',
            docs: 'docs',
            variables: 'variables'
          }
        },
        action_bar: {
          0: '',
          mapping: {
            style: 'style',
            icons: 'icons',
            actions: 'actions',
            docs: 'docs',
            hardcons: 'hardcons',
            prefs: 'prefs',
            variables: 'variables',
            data: 'data'
          }
        },
        net_helper: {
          0: ''
        }
      },
      drive: {
        'style/': {
          'tab_group.css': {
            $ref: 'style/tab_group.css'
          }
        }
      }
    }
  }
}
