// action_bar/action_bar.js
const STATE = require('STATE')
const statedb = STATE(__filename)
const { get } = statedb(fallback_module)

const quick_actions = require('quick_actions')
const actions = require('actions')
const steps_wizard = require('steps_wizard')

module.exports = action_bar

async function action_bar (opts, protocol) {
  const { id, sdb } = await get(opts.sid)
  const { drive } = sdb
  const ids = opts.ids
  if (!ids || !ids.up) {
    throw new Error(`Component ${__filename} requires ids.up to be provided`)
  }
  const by = id
  const to = ids.up

  const on = {
    style: inject,
    icons: iconject
  }

  const el = document.createElement('div')
  const shadow = el.attachShadow({ mode: 'closed' })

  shadow.innerHTML = `
  <div class="container">
    <div class="actions">
      <actions></actions>
    </div>
    <div class="steps-wizard">
      <steps-wizard></steps-wizard>
    </div>
    <div class="action-bar-container main">
      <div class="command-history">
        <button class="icon-btn"></button>
      </div>
      <div class="quick-actions">
        <quick-actions></quick-actions>
      </div>
    </div>
  </div>
  <style>
  </style>`
  const style = shadow.querySelector('style')
  const history_icon = shadow.querySelector('.icon-btn')
  const quick_placeholder = shadow.querySelector('quick-actions')
  const actions_placeholder = shadow.querySelector('actions')
  const steps_wizard_placeholder = shadow.querySelector('steps-wizard')

  let console_icon = {}
  const subs = await sdb.watch(onbatch)

  const _ = {
    up: null,
    send_quick_actions: null,
    send_actions: null,
    send_steps_wizard: null
  }
  let actions_data = null
  let selected_action = null

  if (protocol) {
    const send = protocol(msg => onmessage(msg))
    _.up = send
  }

  let mid = 0

  const quick_actions_sid = subs[0].sid
  const actions_sid = subs[1].sid
  const steps_wizard_sid = subs[2].sid

  history_icon.innerHTML = console_icon
  history_icon.onclick = onhistory
  const element = protocol ? await quick_actions({ ...subs[0], ids: { up: id } }, quick_actions_protocol) : await quick_actions({ ...subs[0], ids: { up: id } })
  quick_placeholder.replaceWith(element)

  const actions_el = protocol ? await actions({ ...subs[1], ids: { up: id } }, actions_protocol) : await actions({ ...subs[1], ids: { up: id } })
  actions_el.classList.add('hide')
  actions_placeholder.replaceWith(actions_el)

  const steps_wizard_el = protocol ? await steps_wizard({ ...subs[2], ids: { up: id } }, steps_wizard_protocol) : await steps_wizard({ ...subs[2], ids: { up: id } })
  steps_wizard_el.classList.add('hide')
  steps_wizard_placeholder.replaceWith(steps_wizard_el)

  const parent_handler = {
    load_actions,
    selected_action: parent__selected_action,
    show_submit_btn,
    hide_submit_btn,
    form_data,
    update_actions_for_app,
    update_quick_actions_for_app
  }

  return el

  async function onbatch (batch) {
    for (const { type, paths } of batch) {
      const data = await Promise.all(paths.map(path => drive.get(path).then(file => file.raw)))
      const func = on[type] || fail
      func(data, type)
    }
  }
  function fail (data, type) { console.warn('Unknown message type:', type, data) }

  function inject (data) {
    style.innerHTML = data.join('\n')
  }

  function iconject (data) {
    console_icon = data[0]
  }
  function onhistory () {
    const head = [by, to, mid++]
    const refs = {}
    _.up({ head, refs, type: 'console_history_toggle', data: null })
    const head2 = [by, to, mid++]
    _.up({ head: head2, refs, type: 'ui_focus', data: 'command_history' })
  }

  // --- Toggle Views ---
  function toggle_view (el, show) {
    el.classList.toggle('hide', !show)
  }

  function actions_toggle_view (display) {
    toggle_view(actions_el, display === 'block')
  }

  function steps_toggle_view (display) {
    toggle_view(steps_wizard_el, display === 'block')
  }

  // -------------------------------
  // Protocol: actions
  // -------------------------------

  function actions_protocol (send) {
    _.send_actions = send

    const actions_handlers = {
      selected_action: actions__selected_action
    }

    return function on (msg) {
      if(msg.type === "ui_focus"){
        _.up(msg)
        return
      }
      const { type } = msg
      const handler = actions_handlers[type] || fail
      handler(msg)
    }
  }

  function actions__selected_action (msg) {
    const { type, data } = msg
    selected_action = data?.action || null

    const head_to_quick = [by, quick_actions_sid, mid++]
    _.send_quick_actions?.({
      head: head_to_quick,
      refs: msg.head ? { cause: msg.head } : undefined,
      type,
      data: {
        ...data,
        total_steps: actions_data[selected_action]?.length || 0
      }
    })

    const head_to_steps = [by, steps_wizard_sid, mid++]
    _.send_steps_wizard?.({ head: head_to_steps, refs: msg.head ? { cause: msg.head } : undefined, type: 'init_data', data: actions_data[selected_action] })

    steps_toggle_view('block')

    if (actions_data[selected_action]?.length > 0) {
      const first_step = actions_data[selected_action][0]
      const head = [by, to, mid++]
      const refs = msg.head ? { cause: msg.head } : undefined
      _.up?.({ head, refs, type: 'render_form', data: first_step })
    }

    if (actions_data[selected_action][actions_data[selected_action].length - 1]?.is_completed) {
      const head_to_quick_submit = [by, quick_actions_sid, mid++]
      _?.send_quick_actions({ head: head_to_quick_submit, refs: msg.head ? { cause: msg.head } : undefined, type: 'show_submit_btn' })
    }

    _.up?.({ head: msg.head, refs: msg.refs, type, data: selected_action })
    actions_toggle_view('none')
  }

  // -------------------------------
  // Protocol: quick actions
  // -------------------------------

  function quick_actions_protocol (send) {
    _.send_quick_actions = send

    const quick_handlers = {
      display_actions: quick_actions__display_actions,
      action_submitted: quick_actions__action_submitted,
      filter_actions,
      update_quick_actions_input
    }

    return on
    function on (msg) {
      const { type } = msg
      const handler = quick_handlers[type] || fail
      handler(msg)
    }
  }

  function quick_actions__display_actions (msg) {
    const { data } = msg
    actions_toggle_view(data)
    if (data === 'none') {
      steps_toggle_view('none')
      const head = [by, to, mid++]
      const refs = msg.head ? { cause: msg.head } : undefined
      _.up?.({ head, refs, type: 'clean_up', data: selected_action })
    }
  }

  function quick_actions__action_submitted (msg) {
    const result = JSON.stringify(actions_data[selected_action].map(step => step.data), null, 2)
    const head_to_quick = [by, quick_actions_sid, mid++]
    const refs_to_quick = msg.head ? { cause: msg.head } : undefined
    _.send_quick_actions?.({ head: head_to_quick, refs: refs_to_quick, type: 'deactivate_input_field', data: null })
    const head = [by, to, mid++]
    const refs = msg.head ? { cause: msg.head } : undefined
    _.up?.({ head, refs, type: 'action_submitted', data: { result, selected_action } })
  }

  // -------------------------------
  // Protocol: steps wizard
  // -------------------------------

  function steps_wizard_protocol (send) {
    _.send_steps_wizard = send

    const steps_handlers = {
      step_clicked: steps_wizard__step_clicked
    }

    return function on (msg) {
      const { type } = msg
      const handler = steps_handlers[type]
      handler(msg)
    }
  }

  function steps_wizard__step_clicked (msg) {
    const { data } = msg
    const head_to_quick = [by, quick_actions_sid, mid++]
    _.send_quick_actions?.({ head: head_to_quick, type: 'update_current_step', data })
    const head = [by, to, mid++]
    const refs = msg.head ? { cause: msg.head } : {}
    _.up?.({ head, refs, type: 'render_form', data })
    const head2 = [by, to, mid++]
    _.up({ head: head2, refs, type: 'ui_focus', data: 'wizard hat' })
  }

  function onmessage (msg) {
    const { type } = msg
    parent_handler[type]?.(msg)
  }

  function load_actions (msg) {
    const { data, type } = msg
    actions_data = data
    const head_to_actions = [by, actions_sid, mid++]
    _.send_actions?.({ head: head_to_actions, type, data })
  }
  function parent__selected_action (msg) {
    const head_to_quick = [by, quick_actions_sid, mid++]
    _.send_quick_actions?.({ head: head_to_quick, ...msg })
  }
  function show_submit_btn (msg) {
    const head_to_quick = [by, quick_actions_sid, mid++]
    _.send_quick_actions?.({ head: head_to_quick, type: 'show_submit_btn' })
  }
  function hide_submit_btn (msg) {
    const head_to_quick = [by, quick_actions_sid, mid++]
    _.send_quick_actions?.({ head: head_to_quick, type: 'hide_submit_btn' })
  }
  function form_data (msg) {
    const head_to_steps = [by, steps_wizard_sid, mid++]
    _.send_steps_wizard?.({ head: head_to_steps, type: 'init_data', data: actions_data[selected_action] })
  }
  
  function update_actions_for_app (msg) {
    const { data, type } = msg
    console.log('Action Bar: Updating actions for focused app:', data?.focused_app)
    
    // Forward update_actions_for_app to actions and quick_actions submodules
    const refs = msg.head ? { cause: msg.head } : {}
    
    const head_to_actions = [by, actions_sid, mid++]
    _.send_actions?.({ head: head_to_actions, refs, type, data })
    
    const head_to_quick = [by, quick_actions_sid, mid++]
    _.send_quick_actions?.({ head: head_to_quick, refs, type, data })
  }

  function update_quick_actions_for_app (msg) {
    const { data, type } = msg
    const head_to_quick_actions = [by, quick_actions_sid, mid++]
    const refs = msg.head ? { cause: msg.head } : {}
    _.send_quick_actions?.({ head: head_to_quick_actions, refs, type, data })
  }

  function filter_actions (msg) {
    const { data, type } = msg
    const head_to_actions = [by, actions_sid, mid++]
    _.send_actions?.({ head: head_to_actions, cause: msg.head ? { cause: msg.head } : {}, type, data })
  }

  function update_quick_actions_input (msg) {
    const { data } = msg
    const head_to_quick = [by, quick_actions_sid, mid++]
    _.send_quick_actions?.({ 
      head: head_to_quick, 
      type: 'update_input_command', 
      data,
      refs: msg.head ? { cause: msg.head } : {}
    })
  }
}

function fallback_module () {
  return {
    api: fallback_instance,
    _: {
      quick_actions: { $: '' },
      actions: { $: '' },
      steps_wizard: { $: '' }
    }
  }
  function fallback_instance () {
    return {
      _: {
        quick_actions: {
          0: '',
          mapping: {
            style: 'style',
            icons: 'icons',
            actions: 'actions',
            hardcons: 'hardcons',
            prefs: 'prefs'
          }
        },
        actions: {
          0: '',
          mapping: {
            style: 'style',
            icons: 'icons',
            actions: 'actions',
            hardcons: 'hardcons'
          }
        },
        steps_wizard: {
          0: '',
          mapping: {
            style: 'style',
            variables: 'variables'
          }
        }
      },
      drive: {
        'icons/': {
          'console.svg': {
            $ref: 'console.svg'
          }
        },
        'style/': {
          'theme.css': {
            raw: `
              .container {
                display: flex;
                flex-direction: column;
              }
              .action-bar-container {
                display: flex;
                flex-direction: row;
                flex-wrap: nowrap;
                align-items: center;
                background: #131315;
                padding: 8px;
                gap: 12px;
              }
              .command-history {
                display: flex;
                align-items: center;
              }
              .quick-actions {
                display: flex;
                flex: auto;
                flex-direction: row;
                flex-wrap: nowrap;
                align-items: center;
                min-width: 300px;
              }
              .hide {
                display: none;
              }
              
              .icon-btn {
                display: flex;
                min-width: 32px;
                height: 32px;
                border: none;
                background: transparent;
                cursor: pointer;
                flex-direction: row;
                justify-content: center;
                align-items: center;
                padding: 6px;
                border-radius: 6px;
                color: #a6a6a6;
              }
              .icon-btn:hover {
                background: rgba(255, 255, 255, 0.1);
              }
              svg {
                width: 20px;
                height: 20px;
              }
            `
          }
        },
        'actions/': {},
        'hardcons/': {},
        'prefs/': {},
        'variables/': {}
      }
    }
  }
}
