// tabsbar/tabsbar.js
const state = require('STATE')
const state_db = state(__filename)
const { get } = state_db(fallback_module)
const DOCS = require('DOCS')

const tabs_component = require('tabs')
const task_manager = require('task_manager')

module.exports = tabsbar

async function tabsbar (opts, protocol) {
  const { id, sdb } = await get(opts.sid)
  const { drive } = sdb
  const ids = opts.ids
  if (!ids || !ids.up) {
    throw new Error(`Component ${__filename} requires ids.up to be provided`)
  }
  const by = id
  const to = ids.up

  const on = {
    style: inject,
    icons: inject_icons
  }

  let dricons = {}
  let mid = 0
  let docs_toggle_active = false
  const el = document.createElement('div')
  const shadow = el.attachShadow({ mode: 'closed' })

  let send = null
  let _ = null
  if (protocol) {
    send = protocol(msg => onmessage(msg))
    _ = { up: send, tabs: null, task_manager: null }
  }

  shadow.innerHTML = `
  <div class="tabs-bar-container main">
  <button class="hat-btn"></button>
  <tabs></tabs>
  <task-manager></task-manager>
  <button class="bar-btn"></button>
  </div>
  <style>
  </style>`
  const style = shadow.querySelector('style')
  const hat_btn = shadow.querySelector('.hat-btn')
  const bar_btn = shadow.querySelector('.bar-btn')

  const subs = await sdb.watch(onbatch)
  const docs = DOCS(__filename)(opts.sid)
  if (dricons[0]) {
    const parser = new DOMParser()
    const doc = parser.parseFromString(dricons[0], 'image/svg+xml')
    const svgElem = doc.documentElement
    hat_btn.replaceChildren(svgElem)
    hat_btn.onclick = docs.wrap(hat_click, async () => {
      const doc_file = await drive.get('docs/README.md')
      return doc_file?.raw || 'No documentation available'
    })
  }
  if (dricons[2]) {
    const parser = new DOMParser()
    const doc = parser.parseFromString(dricons[2], 'image/svg+xml')
    const svgElem = doc.documentElement
    bar_btn.replaceChildren(svgElem)
    bar_btn.onclick = () => {
      docs_toggle_active = !docs_toggle_active
      const head = [by, to, mid++]
      const head_mgr = [by, to, mid++]
      const refs = {}
      // Send message to root module to set docs mode
      _.up?.({ head, refs, type: 'set_docs_mode', data: { active: docs_toggle_active } })
      // Also send docs_toggle notification for UI updates
      _.up?.({ head: [by, to, mid++], refs, type: 'docs_toggle', data: { active: docs_toggle_active } })
      bar_btn.classList.toggle('active', docs_toggle_active)
      _.task_manager({ head_mgr, refs, type: 'docs_toggle', data: { active: docs_toggle_active } })
    }
  }
  const tabs = protocol ? await tabs_component({ ...subs[0], ids: { up: id } }, tabs_protocol) : await tabs_component({ ...subs[0], ids: { up: id } })
  tabs.classList.add('tabs-bar')
  shadow.querySelector('tabs').replaceWith(tabs)

  const task_mgr = protocol ? await task_manager({ ...subs[1], ids: { up: id } }, task_manager_protocol) : await task_manager({ ...subs[1], ids: { up: id } })
  task_mgr.classList.add('bar-btn')
  shadow.querySelector('task-manager').replaceWith(task_mgr)

  return el
  async function hat_click () {
    const head = [by, to, mid++]
    const refs = {}
    _.up?.({ head, refs, type: 'ui_focus', data: 'wizard_hat' })
  }
  function onmessage (msg) {
    const { type } = msg
    switch (type) {
    case 'docs_toggle':
      // Broadcast to subcomponents
      _.tabs?.(msg)
      break
    default:
        // Handle other message types
    }
  }

  function tabs_protocol (send) {
    _.tabs = send
    return on
    function on (msg) {
      _.up(msg)
    }
  }

  function task_manager_protocol (send) {
    _.task_manager = send
    return on
    function on (msg) {
      _.up(msg)
    }
  }

  async function onbatch (batch) {
    for (const { type, paths } of batch) {
      const data = await Promise.all(paths.map(path => drive.get(path).then(file => file.raw)))
      const func = on[type] || fail
      func(data, type)
    }
  }
  function fail (data, type) { console.warn('invalid message', { cause: { data, type } }) }

  function inject (data) {
    style.innerHTML = data.join('\n')
  }

  function inject_icons (data) {
    dricons = data
  }
}

function fallback_module () {
  return {
    api: fallback_instance,
    _: {
      tabs: {
        $: ''
      },
      task_manager: {
        $: ''
      },
      DOCS: {
        $: ''
      }
    }
  }

  function fallback_instance () {
    return {
      _: {
        tabs: {
          0: '',
          mapping: {
            icons: 'icons',
            variables: 'variables',
            scroll: 'scroll',
            style: 'style',
            docs: 'docs'
          }
        },
        task_manager: {
          0: '',
          mapping: {
            count: 'count',
            style: 'style',
            docs: 'docs'
          }
        },
        DOCS: {
          0: ''
        }
      },
      drive: {
        'style/': {
          'theme.css': {
            raw: `
              .tabs-bar-container {
                display: flex;
                flex: inherit;
                flex-direction: row;
                flex-wrap: nowrap;
                align-items: stretch;
              }
              .tabs-bar {
                display: flex;
                flex: auto;
                flex-direction: row;
                flex-wrap: nowrap;
                align-items: stretch;
                width: 300px;
              }
              .hat-btn, .bar-btn {
                display: flex;
                min-width: 32px;
                border: none;
                background: #131315;
                cursor: pointer;
                flex-direction: row;
                justify-content: center;
                align-items: center;
              }
              .bar-btn.active {
                background: #2d4a6d;
              }
            `
          }
        },
        'icons/': {
          '1.svg': {
            $ref: 'hat.svg'
          },
          '2.svg': {
            $ref: 'hat.svg'
          },
          '3.svg': {
            $ref: 'docs.svg'
          }
        },
        'docs/': {
          'README.md': {
            $ref: 'README.md'
          }
        }
      }
    }
  }
}
