// DOCS/index.js
// --- Main Export ---
// Usage: const docs = DOCS(__filename)(opts.sid)
//        docs.wrap(handler, docContent)
// Admin: Only first caller (root module) gets admin API

module.exports = function DOCS (filename) {
  return function (sid) { return create_context(filename, sid) }
}

const scope = typeof window !== 'undefined' ? window : global

if (!scope.__DOCS_GLOBAL_STATE__) {
  scope.__DOCS_GLOBAL_STATE__ = {
    docs_mode_active: false,
    docs_mode_listeners: [],
    doc_display_callback: null,
    action_registry: new Map()
  }
}

const state = scope.__DOCS_GLOBAL_STATE__

// --- Static Methods (called as DOCS.method()) ---
// Exported via DOCS admin API (only available to first caller)
function set_docs_mode (active) {
  state.docs_mode_active = active
  state.docs_mode_listeners.forEach(listener => listener(active))
}

function get_docs_mode () { return state.docs_mode_active }

function on_docs_mode_change (listener) {
  state.docs_mode_listeners.push(listener)
  return unsubscribe_docs_mode_change

  function unsubscribe_docs_mode_change () { state.docs_mode_listeners = state.docs_mode_listeners.filter(l => l !== listener) }
}

function set_doc_display_handler (callback) { state.doc_display_callback = callback }

function get_actions (sid) {
  const actions = state.action_registry.get(sid) || []
  if (actions.length === 0) throw new Error('DOCS: No actions registered for SID ' + sid)
  return actions
}

function list_registered () { return Array.from(state.action_registry.keys()) }

// --- Internal Helpers ---

function verify_actions (actions) {
  if (!Array.isArray(actions)) throw new Error('DOCS: Actions must be array')
  actions.forEach(validate_action)

  function validate_action (action, i) {
    if (!action.name || typeof action.name !== 'string') throw new Error(`DOCS: Action[${i}] Invalid 'name'`)
    if (!action.info || typeof action.info !== 'string') throw new Error(`DOCS: Action[${i}] Invalid 'info'`)
    if (!action.icon || typeof action.icon !== 'string') throw new Error(`DOCS: Action[${i}] Invalid 'icon'`)
    if (!action.status || typeof action.status !== 'object') throw new Error(`DOCS: Action[${i}] Invalid 'status'`)
    if (!action.steps || !Array.isArray(action.steps)) throw new Error(`DOCS: Action[${i}] Invalid 'steps'`)
  }
}

async function display_doc (content, sid) {
  let resolved_content = content
  if (typeof content === 'function') {
    resolved_content = await content()
  } else if (typeof content.then === 'function') {
    resolved_content = await content
  }

  if (state.doc_display_callback) {
    state.doc_display_callback({ content: resolved_content || 'No documentation available', sid })
  }
}

function create_sys_api (meta) {
  const messages = create_message_api(meta)
  const drive = create_drive_api(meta)
  const sdb = create_sdb_api(meta)

  return {
    is_docs_mode: () => state.docs_mode_active,
    get_doc: () => meta.doc || 'No documentation available',
    get_meta: () => ({ ...meta }),
    show_doc: () => display_doc(meta.doc || 'No documentation available', meta.sid),
    show_action_info: action => show_action_info(meta.sid, action),
    get_actions: () => state.action_registry.get(meta.sid) || [],
    get_action: name => get_action(meta.sid, name),
    send: (channel, type, refs, data) => send_message(meta, channel, type, refs, data),
    _: messages,
    drive,
    sdb,
    get_drive: () => drive,
    get_sdb: () => sdb
  }
}

function get_action (sid, name) {
  const actions = state.action_registry.get(sid) || []
  return actions.find(action => action.name === name) || null
}

function get_sys_resources (meta) {
  return meta.resources || {}
}

function create_message_api (meta) {
  return new Proxy({}, {
    get: function get_channel (_, channel) {
      return function send_channel_message (type, refs, data) {
        return send_message(meta, channel, type, refs, data)
      }
    }
  })
}

function send_message (meta, channel, type, refs, data) {
  if (state.docs_mode_active) return false
  const sender = get_sys_resources(meta)._
  if (!sender || typeof sender[channel] !== 'function') throw new Error('DOCS: sys._.' + channel + ' is unavailable')
  return sender[channel](type, refs, data)
}

function create_drive_api (meta) {
  return {
    get: path => require_drive(meta).get(path),
    put: (path, data) => state.docs_mode_active ? false : require_drive(meta).put(path, data)
  }
}

function require_drive (meta) {
  const drive = get_sys_resources(meta).drive
  if (!drive) throw new Error('DOCS: sys.drive is unavailable')
  return drive
}

function create_sdb_api (meta) {
  return {
    watch: handler => state.docs_mode_active ? noop_unsubscribe : require_sdb(meta).watch(handler),
    drive: create_drive_api(meta)
  }
}

function require_sdb (meta) {
  const sdb = get_sys_resources(meta).sdb
  if (!sdb) throw new Error('DOCS: sys.sdb is unavailable')
  return sdb
}

function noop_unsubscribe () {}

// --- Instance Methods (called as docs.method()) ---

function wrap (handler, meta = {}, make_sys = create_sys_api) {
  meta = { ...meta, doc: get_handler_doc(handler, meta.doc) }
  const sys = make_sys(meta)

  return async function wrapped_handler (event) {
    if (sys.is_docs_mode()) {
      if (event.preventDefault) {
        event.preventDefault()
        event.stopPropagation()
      }
      sys.show_doc()
      return
    }
    return handler.call(this, event, sys)
  }

  function get_handler_doc (handler, doc) {
    if (doc !== undefined) return doc
    if (handler.docs !== undefined) return handler.docs
    if (handler.info !== undefined) return handler.info
    return doc
  }
}

function wrap_isolated (handler_string, meta = {}) {
  try {
    const params = 'meta, make_sys'
    const source = `(${wrap.toString()})(${handler_string}, ${params})`
    // eslint-disable-next-line no-new-func
    const isolated_fn = new Function(params, source)(meta, create_sys_api)
    return isolated_fn
  } catch (err) {
    console.error('handler function is not allowed to access closure scope', err)
    return wrap(() => {}, meta)
  }
}

function hook (dom, meta = {}) {
  if (!dom) return dom

  const proto = Object.getPrototypeOf(Object.getPrototypeOf(dom))
  if (!proto) return dom

  Object.keys(proto).forEach(hook_event_handler)

  function hook_event_handler (key) {
    if (key.startsWith('on') && typeof dom[key] === 'function') {
      const original = dom[key]
      dom[key] = wrap(original, { ...meta, event_type: key })
    }
  }

  return dom
}

// --- Context Factory (creates instance with component scope) ---

function register_actions (sid, actions) {
  verify_actions(actions)
  state.action_registry.set(sid, actions)
}

function show_action_info (sid, action) {
  if (!state.docs_mode_active) return false
  if (!action || typeof action.info !== 'string') return false
  display_doc(action.info, sid)
  return true
}

let admin = true
function create_context (filename, sid) {
  const resources = {}
  const api = {
    wrap: wrap_with_component,
    wrap_isolated: wrap_isolated_with_component,
    hook: hook_with_component,
    get_docs_mode,
    on_docs_mode_change,
    register_actions: register_component_actions,
    show_action_info: show_component_action_info,
    set_sys
  }
  const admin_api = {
    set_docs_mode,
    set_doc_display_handler,
    get_actions,
    list_registered
  }
  const context = admin ? (admin = false, Object.assign({ admin: admin_api }, api)) : api
  return context

  function wrap_with_component (handler, doc) { return wrap(handler, { doc, sid, component: filename, resources }) }
  function wrap_isolated_with_component (handler_string, doc) { return wrap_isolated(handler_string, { doc, sid, component: filename, resources }) }
  function hook_with_component (dom, doc) { return hook(dom, { doc, sid, component: filename, resources }) }
  function register_component_actions (actions) { return register_actions(sid, actions) }
  function show_component_action_info (action) { return show_action_info(sid, action) }
  function set_sys (api) { Object.assign(resources, api) }
}
