// net_helper/net_helper.js
module.exports = net

function net (id) {
  const label = `[${id}]`
  const on = {}
  const _ = {}
  const sub = {}
  const hub = {}

  return { io: { invite, accept, on }, _ }

  function forward (to, msg) {
    if (typeof to === 'string' && to === id) {
      const ups = [...new Set(Object.keys(hub).map(get_hub_tx))]
      for (const tx of ups) tx(msg)
        return
    }
    for (const sub_id of Object.keys(sub)) {
      if (typeof to === 'string' && to === sub_id) {
        sub[sub_id].tx(msg)
        return
      }
    }
    throw new Error(`${label} unknown recipient "${to}"`)

    function get_hub_tx (hub_id) { return hub[hub_id].tx }
  }
  
  function invite (name, ids) {
    if (!on[name]) throw new Error(`${label} no protocol handler for "${name}"`)
    return Object.assign(connect, { ids })

    function connect (tx) {
      const rx = router(sub)
      add(name, tx, tx.id, rx, sub)
      return rx
    }
  }

  function accept (invite) {
    const rx = router(hub)
    const tx = invite(Object.assign(rx, { id }))
    const ids = invite.ids || {}
    for (const [name, to] of Object.entries(ids)) {
      if (hub[to]) throw new Error(`${label} already connected to "${to}"`)
      if (!on[name]) throw new Error(`${label} no "${name}" protocol for "${to}"`)
      add(name, tx, to, rx, hub)
    }
  }

  function router (connections) {
    return function rx (msg) {
      if (!msg || !Array.isArray(msg.head)) throw new Error(`${label} invalid message`)
      const [by, to] = msg.head
      if (to !== id) return forward(to, msg)
      if (!connections[by]) throw new Error(`${label} unknown sender "${by}"`)
      const { name } = connections[by].state
      if (!on[name]) throw new Error(`${label} no "${name}" protocol for "${to}"`)
      on[name](msg)
    }
  }

  function add (name, tx, to, rx, connections) {
    const state = { name, to, mid: 0 }
    const send = create_send(tx, state)
    Object.defineProperties(send, {
      channel: {
        enumerable: true,
        get: get_channel
      },
      to: {
        enumerable: true,
        get: get_to
      },
      mid: {
        enumerable: true,
        get: get_mid
      },
      state: {
        enumerable: false,
        value: state
      }
    })
    _[name] = send
    connections[to] = { rx, tx, state }

    function get_channel () { return state.name }
    function get_to () { return state.to }
    function get_mid () { return state.mid }
  }

  function create_send (tx, state) {
    return function send (type, data = [], refs = {}, meta) {
      const msg = normalize_message(type, data, refs, meta, state)
      tx(msg)
    }
  }

  function normalize_message (type, data, refs, meta, state) {
    if (type && typeof type === 'object') {
      const msg = { ...type }
      const previous_head = msg.head
      msg.head = [id, state.to, state.mid++]
      if (previous_head) {
        const next_refs = msg.refs ? { ...msg.refs } : {}
        if (!next_refs.cause) next_refs.cause = previous_head
        msg.refs = next_refs
      }
      return msg
    }
    const msg = { head: [id, state.to, state.mid++], refs, type, data }
    if (meta !== undefined) msg.meta = meta
    return msg
  }
}
