// control_unit/control_unit.js
const STATE = require('STATE')
const statedb = STATE(__filename)
const { get } = statedb(fallback_module)

module.exports = control_unit

async function control_unit (opts, protocol) {
  const { id, sdb } = await get(opts.sid)
  const { drive } = sdb
  const ids = opts.ids
  if (!ids || !ids.up) {
    throw new Error(`Component ${__filename} requires ids.up to be provided`)
  }
  const by = id
  const to = ids.up

  let send = null
  let _ = null
  let mid = 0

  if (protocol) {
    send = protocol(msg => onmessage(msg))
    _ = { up: send }
  }
  
  await sdb.watch(() => {})
  const welcome_actions = await drive.get('actions/commands.json')
  const welcome_commands = JSON.parse(welcome_actions.raw)
  const refs = {}
  const head = [by, to, mid++]
  _.up({ head, refs, type: 'welcome_actions', data: { type: "welcome", actions: welcome_commands }})
  return 0

  async function onmessage (msg) {
    const { type, data } = msg
    if (type === 'focused_app_changed') {
      if (_.up) {
        const focused_app = data?.type
        let actions_data = null
        let quick_actions_data = null
        
        if (focused_app) {
          const component_actions = await get_component_actions(data)
          actions_data = component_actions.actions
          quick_actions_data = component_actions.quick_actions
        }
        
        const message_data = {
          ...data,
          temp_actions: actions_data
        }
        const head = [by, to, mid++]
        const refs = msg.head ? { cause: msg.head } : {}
        _.up({ head, refs, type: 'update_actions_for_app', data: message_data })

        const quick_actions_message_data = {
          ...data,
          temp_quick_actions: quick_actions_data
        }
        const quick_actions_head = [by, to, mid++]
        const quick_actions_refs = msg.head ? { cause: msg.head } : {}
        _.up({ head: quick_actions_head, refs: quick_actions_refs, type: 'update_quick_actions_for_app', data: quick_actions_message_data })
      }
    }
  }

  async function get_component_actions (data) {
    const { actions } = data
    const result_actions = []
    const result_quick_actions = []
    let temp_actions = {}
    let temp_quick_actions = {}
    actions.forEach(element => {
      temp_actions = {}
      temp_actions.action = element.name
      temp_actions.icon = element.icon
      temp_actions.pinned = element.status.pinned
      temp_actions.default = element.status.default
      result_actions.push(temp_actions)

      temp_quick_actions = {}
      temp_quick_actions.name = element.name
      temp_quick_actions.icon = element.icon
      result_quick_actions.push(temp_quick_actions)
    })
      return {
      actions: result_actions,
      quick_actions: result_quick_actions
    }
  }
}

function fallback_module () {
  return {
    api: fallback_instance
  }
  function fallback_instance () {
    return {
      drive: {
        'actions/': {
          'commands.json': {
            raw: JSON.stringify([
              {
                name: 'New',
                icon: 'file',
                status: {
                  pinned: true,
                  default: true
                },
                steps: [
                  { name: 'Enter File Name', type: 'mandatory', is_completed: false, component: 'form_input', status: 'default', data: '' },
                  { name: 'Choose Location', type: 'mandatory', is_completed: false, component: 'form_input', status: 'default', data: '' }
                ]
              },
              {
                name: 'Settings',
                icon: 'folder',
                status: {
                  pinned: false,
                  default: true
                },
                steps: [
                  { name: 'Configure Settings', type: 'optional', is_completed: false, component: 'form_input', status: 'default', data: '' }
                ]
              },
              {
                name: 'Help',
                icon: 'save',
                status: {
                  pinned: true,
                  default: false
                },
                steps: [
                  { name: 'View Documentation', type: 'optional', is_completed: false, component: 'form_input', status: 'default', data: '' }
                ]
              },
              {
                name: 'About',
                icon: 'gear',
                status: {
                  pinned: false,
                  default: true
                },
                steps: [
                  { name: 'View Information', type: 'optional', is_completed: false, component: 'form_input', status: 'default', data: '' }
                ]
              },
              {
                name: 'Exit',
                icon: 'help',
                status: {
                  pinned: false,
                  default: false
                },
                steps: [
                  { name: 'Confirm Exit', type: 'mandatory', is_completed: false, component: 'form_input', status: 'default', data: '' }
                ]
              }
            ])
          }
        },
        'docs/': {
          'README.md': {
            $ref: 'README.md'
          }
        }
      }
    }
  }
}
