// tabs/tabs.js
const STATE = require('STATE')
const statedb = STATE(__filename)
const { get } = statedb(fallback_module)
const DOCS = require('DOCS')
const net = require('net_helper')

module.exports = component

async function component (opts, invite) {
  const { id, sdb } = await get(opts.sid)
  const { drive } = sdb

  const on = {
    variables: onvariables,
    style: inject,
    icons: iconject,
    scroll: onscroll
  }
  const div = document.createElement('div')
  const shadow = div.attachShadow({ mode: 'closed' })
  shadow.innerHTML = `
  <div class="tab-entries main"></div>`
  const entries = shadow.querySelector('.tab-entries')
  const sheet = new CSSStyleSheet()
  const focus_sheet = new CSSStyleSheet()
  focus_sheet.replaceSync(`
    /* Focus highlighting for tabs */
    .tab-entries.tile-focused .tabsbtn.active {
      background: rgba(103, 195, 255, 0.25);
      border-color: transparent;
    }
    .tab-entries.tile-inactive .tabsbtn.active {
      background: transparent;
      border: 1px solid rgba(103, 195, 255, 0.5);
    }
    /* Tab group inline styles */
    .tab-group-inline {
      display: inline-flex;
      align-items: stretch;
      height: 100%;
    }
    /* Group tab and child tabs share underline when expanded */
    .tab-group-inline.expanded {
      border-bottom: 2px solid rgba(103, 195, 255, 0.8);
    }
    .tab-group-inline .tab-group-tab {
      cursor: pointer;
    }
    .tab-group-inline .tab-group-tab .icon {
      display: inline-flex;
      align-items: center;
      color: rgba(103, 195, 255, 0.8);
    }
    .tab-group-inline .tab-group-tab .icon svg {
      width: 12px;
      height: 12px;
    }
    /* Child tabs hidden by default, shown inline when expanded */
    .tab-group-children {
      display: none;
      align-items: stretch;
    }
    .tab-group-inline.expanded .tab-group-children {
      display: inline-flex;
    }
    .tab-group-child {
      cursor: pointer;
    }
    .tab-group-child .icon {
      display: inline-flex;
      align-items: center;
      color: rgba(103, 195, 255, 0.6);
    }
    .tab-group-child .icon svg {
      width: 12px;
      height: 12px;
    }
  `)
  shadow.adoptedStyleSheets = [sheet, focus_sheet]

  let init = false
  let variables = []
  let dricons = []
  let active = null
  let tab_group_el = null
  let tab_group_expanded = false
  let tile_is_focused = false
  const default_tabs = {}
  const link_tabs = {}
  const ARROW_LEFT_SVG = `<svg width="12" height="12" viewBox="0 0 12 12" fill="none" xmlns="http://www.w3.org/2000/svg"><path d="M10.875 6C10.875 6.09946 10.8355 6.19484 10.7652 6.26517C10.6948 6.33549 10.5995 6.375 10.5 6.375H4.28016L7.01531 9.10969C7.05015 9.14453 7.07779 9.18589 7.09665 9.23141C7.1155 9.27694 7.12521 9.32573 7.12521 9.375C7.12521 9.42427 7.1155 9.47306 7.09665 9.51859C7.07779 9.56411 7.05015 9.60547 7.01531 9.64031C6.98047 9.67515 6.93911 9.70279 6.89359 9.72165C6.84806 9.7405 6.79927 9.75021 6.75 9.75021C6.70073 9.75021 6.65194 9.7405 6.60641 9.72165C6.56089 9.70279 6.51953 9.67515 6.48469 9.64031L3.10969 6.26531C3.07482 6.23049 3.04716 6.18913 3.02829 6.1436C3.00942 6.09808 2.9997 6.04928 2.9997 6C2.9997 5.95072 3.00942 5.90192 3.02829 5.8564C3.04716 5.81087 3.07482 5.76951 3.10969 5.73469L6.48469 2.35969C6.51953 2.32485 6.56089 2.29721 6.60641 2.27835C6.65194 2.2595 6.70073 2.24979 6.75 2.24979C6.79927 2.24979 6.84806 2.2595 6.89359 2.27835C6.93911 2.29721 6.98047 2.32485 7.01531 2.35969C7.05015 2.39453 7.07779 2.43589 7.09665 2.48141C7.1155 2.52694 7.12521 2.57573 7.12521 2.625C7.12521 2.67427 7.1155 2.72306 7.09665 2.76859C7.07779 2.81411 7.05015 2.85547 7.01531 2.89031L4.28016 5.625H10.5C10.5995 5.625 10.6948 5.66451 10.7652 5.73484C10.8355 5.80516 10.875 5.90054 10.875 6ZM1.875 1.5C1.77554 1.5 1.68016 1.53951 1.60984 1.60984C1.53951 1.68016 1.5 1.77554 1.5 1.875V10.125C1.5 10.2245 1.53951 10.3198 1.60984 10.3902C1.68016 10.4605 1.77554 10.5 1.875 10.5C1.97446 10.5 2.06984 10.4605 2.14016 10.3902C2.21049 10.3198 2.25 10.2245 2.25 10.125V1.875C2.25 1.77554 2.21049 1.68016 2.14016 1.60984C2.06984 1.53951 1.97446 1.5 1.875 1.5Z" fill="currentColor"/></svg>`
  const ARROW_RIGHT_SVG = `<svg width="12" height="12" viewBox="0 0 12 12" fill="none" xmlns="http://www.w3.org/2000/svg"><path d="M8.89031 5.73469C8.92518 5.76951 8.95284 5.81087 8.97171 5.8564C8.99058 5.90192 9.00029 5.95072 9.00029 6C9.00029 6.04928 8.99058 6.09808 8.97171 6.1436C8.95284 6.18913 8.92518 6.23049 8.89031 6.26531L5.51531 9.64031C5.48047 9.67515 5.43911 9.70279 5.39359 9.72165C5.34806 9.7405 5.29927 9.75021 5.25 9.75021C5.20073 9.75021 5.15194 9.7405 5.10641 9.72165C5.06089 9.70279 5.01953 9.67515 4.98469 9.64031C4.94985 9.60547 4.92221 9.56411 4.90335 9.51859C4.8845 9.47306 4.87479 9.42427 4.87479 9.375C4.87479 9.32573 4.8845 9.27694 4.90335 9.23141C4.92221 9.18589 4.94985 9.14453 4.98469 9.10969L7.71984 6.375H1.5C1.40054 6.375 1.30516 6.33549 1.23483 6.26517C1.16451 6.19484 1.125 6.09946 1.125 6C1.125 5.90054 1.16451 5.80516 1.23483 5.73484C1.30516 5.66451 1.40054 5.625 1.5 5.625H7.71984L4.98469 2.89031C4.91432 2.81995 4.87479 2.72451 4.87479 2.625C4.87479 2.52549 4.91432 2.43005 4.98469 2.35969C5.05505 2.28932 5.15049 2.24979 5.25 2.24979C5.34951 2.24979 5.44495 2.28932 5.51531 2.35969L8.89031 5.73469ZM10.125 1.5C10.0255 1.5 9.93016 1.53951 9.85983 1.60984C9.78951 1.68016 9.75 1.77554 9.75 1.875V10.125C9.75 10.2245 9.78951 10.3198 9.85983 10.3902C9.93016 10.4605 10.0255 10.5 10.125 10.5C10.2245 10.5 10.3198 10.4605 10.3902 10.3902C10.4605 10.3198 10.5 10.2245 10.5 10.125V1.875C10.5 1.77554 10.4605 1.68016 10.3902 1.60984C10.3198 1.53951 10.2245 1.5 10.125 1.5Z" fill="currentColor"/></svg>`
  const docs = DOCS(__filename)(opts.sid)
  const { io, _ } = net(id)

  // Register actions with DOCS system
  const actions_file = await drive.get('actions/commands.json')
  if (actions_file.raw) {
    const actions_data = typeof actions_file.raw === 'string' ? JSON.parse(actions_file.raw) : actions_file.raw
    docs.register_actions(actions_data)
  }

  await sdb.watch(onbatch)
  io.on = {
    up: protocol_from_up()
  }
  if (invite) io.accept(invite)
  if (entries) {
    let is_down = false
    let start_x
    let scroll_start

    function stop_drag_scroll () {
      is_down = false
      entries.classList.remove('grabbing')
      update_scroll_position()
    }

    function move_drag_scroll (pointer_x) {
      if (!is_down) return
      if (entries.scrollWidth <= entries.clientWidth) return stop_drag_scroll()
      entries.scrollLeft = scroll_start - (pointer_x - start_x) * 1.5
    }

    entries.onmousedown = on_entries_mousedown

    function on_entries_mousedown (e) {
      if (entries.scrollWidth <= entries.clientWidth) return
      is_down = true
      entries.classList.add('grabbing')
      start_x = e.pageX - entries.offsetLeft
      scroll_start = entries.scrollLeft
      window.onmousemove = on_window_mousemove
      window.onmouseup = on_window_mouseup
    }

    function on_window_mousemove (e) {
      move_drag_scroll(e.pageX - entries.offsetLeft)
      e.preventDefault()
    }

    function on_window_mouseup () {
      stop_drag_scroll()
      window.onmousemove = null
      window.onmouseup = null
    }

    entries.onmouseleave = stop_drag_scroll
    entries.ontouchstart = on_entries_touchstart

    function on_entries_touchstart (e) {
      if (entries.scrollWidth <= entries.clientWidth) return
      is_down = true
      start_x = e.touches[0].pageX - entries.offsetLeft
      scroll_start = entries.scrollLeft
    }

    ;['ontouchend', 'ontouchcancel'].forEach(bind_touch_end_handler)
    entries.ontouchmove = on_entries_touchmove

    function bind_touch_end_handler (event_name) { entries[event_name] = stop_drag_scroll }
    function on_entries_touchmove (e) {
      move_drag_scroll(e.touches[0].pageX - entries.offsetLeft)
      e.preventDefault()
    }
  }
  return div

  function protocol_from_up () {
    const on = {
      add_link_tab: handle_add_link_tab,
      remove_link_tab: handle_remove_link_tab,
      add_default_tab: handle_add_default_tab,
      show_collapsed_tab_group: handle_show_collapsed_tab_group,
      hide_collapsed_tab_group: handle_hide_collapsed_tab_group,
      tile_focus_changed: handle_tile_focus_changed
    }
    return function onmessage (msg) {
      console.error('tabs: message from up', msg)
      ;(on[msg.type] || onfail)(msg)
    }
    function handle_add_link_tab ({ data }) { add_link_tab(data) }
    function handle_remove_link_tab ({ data }) { remove_link_tab(data) }
    function handle_add_default_tab ({ data }) { add_default_tab(data) }
    function handle_show_collapsed_tab_group ({ data }) { show_collapsed_tab_group(data) }
    function handle_hide_collapsed_tab_group () { hide_collapsed_tab_group() }
    function handle_tile_focus_changed ({ data }) { update_tab_focus_state(data) }
    function onfail (msg) { console.error('tabs: unknown message', msg) }
  }

  function add_default_tab ({ name, program, tile_id }) {
    const tab_id = `tab_${Date.now()}`
    const el = document.createElement('div')
    el.innerHTML = `
    <span class="icon">${dricons[1] || '📄'}</span>
    <span class='name'>${tab_id}</span>
    <span class="name">${name || 'New Tab'}</span>
    <button class="btn">${dricons[0] || '×'}</button>`
    el.className = 'tabsbtn default-tab active'
    const name_el = el.querySelector('.name')
    const close_btn = el.querySelector('.btn')
    default_tabs[tab_id] = { el, name_el, close_btn, name, program }
    if (active) active.classList.remove('active')
    active = el
    name_el.onclick = switch_active
    close_btn.onclick = close_tab
    entries.appendChild(el)
    console.error('tabs: default tab added', tab_id)

    function switch_active () {
      console.error('tabs: default tab clicked', tab_id)
      if (active) active.classList.remove('active')
      el.classList.add('active')
      active = el
      const data = { id: tab_id, name, program }
      _.up('ui_focus', {}, { type: 'tab', sid: opts.sid })
      _.up('tab_name_clicked', {}, data)
    }
    function close_tab (e) {
      e.stopPropagation()
      console.error('tabs: default tab close clicked', tab_id)
      el.remove()
      delete default_tabs[tab_id]
      if (active === el) active = null
      _.up('tab_close_clicked', {}, { id: tab_id, name })
      if (Object.keys(default_tabs).length === 0) {
        _.up('all_tabs_closed', {}, null)
      }
    }
  }

  function add_link_tab ({ tile_id, name, direction }) {
    const link_tab_id = `split_tile_${tile_id}`
    if (link_tabs[link_tab_id]) {
      console.error('tabs: link tab already exists', link_tab_id)
      return
    }
    const el = document.createElement('div')
    el.innerHTML = `
    <span class="icon">⊞</span>
    <span class="name">${name || 'Split ' + direction}</span>`
    el.className = 'tabsbtn link-tab'
    const name_el = el.querySelector('.name')
    link_tabs[link_tab_id] = { el, name_el, tile_id, name, direction }
    el.onclick = on_link_tab_click
    entries.appendChild(el)
    console.error('tabs: link tab added', link_tab_id)

    function on_link_tab_click () {
      console.error('tabs: link tab clicked', link_tab_id)
      _.up('link_tab_clicked', {}, { tile_id, link_tab_id })
    }
  }

  function remove_link_tab ({ tile_id }) {
    const link_tab_id = `split_tile_${tile_id}`
    const tab = link_tabs[link_tab_id]
    if (tab) {
      tab.el.remove()
      delete link_tabs[link_tab_id]
      console.error('tabs: link tab removed', link_tab_id)
    }
  }

  // ---------------------------
  // COLLAPSED TAB GROUP
  // Shows collapsed tile's tabs inline (like Chrome tab groups)
  // ---------------------------

  function show_collapsed_tab_group ({ tiles }) {
    if (tab_group_el) hide_collapsed_tab_group()

    // Container holds the group tab + inline child tabs
    tab_group_el = document.createElement('div')
    tab_group_el.className = 'tab-group-inline'

    // Group tab — styled same as default tabs
    const group_tab = document.createElement('div')
    group_tab.className = 'tabsbtn tab-group-tab'
    const direction = (tiles[0] && tiles[0].direction) || 'right'
    const icon_svg = (direction === 'left' || direction === 'up') ? ARROW_LEFT_SVG : ARROW_RIGHT_SVG
    group_tab.innerHTML = `<span class="icon">${icon_svg}</span><span class="name">Split</span>`
    group_tab.onclick = toggle_tab_group_expand
    tab_group_el.appendChild(group_tab)

    // Child tabs — the actual tabs from the collapsed tile, shown inline
    const children_container = document.createElement('div')
    children_container.className = 'tab-group-children'

    for (const tile of tiles) {
      const tile_tabs = tile.tabs || []
      for (const tab of tile_tabs) {
        const tab_el = document.createElement('div')
        tab_el.className = 'tabsbtn tab-group-child'
        tab_el.innerHTML = `<span class="icon">${dricons[1] || '📄'}</span><span class="name">${tab.name || 'Tab'}</span>`
        tab_el.onclick = on_child_tab_click(tile.tile_id)
        children_container.appendChild(tab_el)
      }
      // If no tabs tracked, show at least one entry
      if (tile_tabs.length === 0) {
        const tab_el = document.createElement('div')
        tab_el.className = 'tabsbtn tab-group-child'
        tab_el.innerHTML = `<span class="icon">${dricons[1] || '📄'}</span><span class="name">New Tab</span>`
        tab_el.onclick = on_child_tab_click(tile.tile_id)
        children_container.appendChild(tab_el)
      }
    }

    tab_group_el.appendChild(children_container)
    entries.insertBefore(tab_group_el, entries.firstChild)
    console.error('tabs: collapsed tab group shown')

    function toggle_tab_group_expand () {
      tab_group_expanded = !tab_group_expanded
      tab_group_el.classList.toggle('expanded', tab_group_expanded)
    }

    function on_child_tab_click (tile_id) {
      return function () {
        console.error('tabs: tab group child clicked, requesting expand', tile_id)
        _.up('tab_group_tile_clicked', {}, { tile_id })
      }
    }
  }

  function hide_collapsed_tab_group () {
    if (tab_group_el) {
      tab_group_el.remove()
      tab_group_el = null
      tab_group_expanded = false
      console.error('tabs: collapsed tab group hidden')
    }
  }

  // ---------------------------
  // FOCUS STATE
  // Updates active tab styling based on tile focus:
  // - Tile in focus: active tab gets highlight background
  // - Tile NOT in focus: active tab gets highlight border only
  // ---------------------------

  function update_tab_focus_state ({ is_focused }) {
    tile_is_focused = is_focused
    entries.classList.toggle('tile-focused', is_focused)
    entries.classList.toggle('tile-inactive', !is_focused)
  }

  async function create_btn ({ name, id }, index) {
    const el = document.createElement('div')
    el.innerHTML = `
    <span class="icon">${dricons[index + 1]}</span>
    <span class='name'>${id}</span>
    <span class="name">${name}</span>
    <button class="btn">${dricons[0]}</button>`

    el.className = 'tabsbtn'
    const name_el = el.querySelector('.name')
    const close_btn = el.querySelector('.btn')

    name_el.draggable = false

    name_el.onclick = docs.wrap(on_tab_name_click, get_doc_content)
    close_btn.onclick = docs.wrap(on_tab_close_click, get_doc_content)

    async function on_tab_name_click () {
      const data = { type: 'tab', sid: opts.sid }
      _.up('ui_focus', {}, data)
      _.up('tab_name_clicked', {}, { id, name })
    }
    async function on_tab_close_click (e) {
      e.stopPropagation()
      const data = { type: 'tab', sid: opts.sid }
      _.up('ui_focus', {}, data)
      _.up('tab_close_clicked', {}, { id, name })
    }
    async function get_doc_content () {
      const doc_file = await drive.get('docs/README.md')
      return doc_file.raw || 'No documentation available'
    }

    entries.appendChild(el)
  }

  async function onbatch (batch) {
    for (const { type, paths } of batch) {
      const data = await Promise.all(paths.map(load_path_raw))
      const func = on[type] || fail
      func(data, type)
    }
    if (!init) {
      if (!opts.ids) variables.forEach(create_btn)
      init = true
    } else {
      // TODO: Here we can handle drive updates
    }

    function load_path_raw (path) { return drive.get(path).then(read_drive_file_raw) }
    function read_drive_file_raw (file) { return file.raw }
  }
  function fail (data, type) { console.warn('invalid message', { cause: { data, type } }) }
  function inject (data) { sheet.replaceSync(data[0]) }

  function onvariables (data) {
    const vars = typeof data[0] === 'string' ? JSON.parse(data[0]) : data[0]
    variables = vars
  }

  function iconject (data) { dricons = data }

  function update_scroll_position () {
    // TODO
  }

  function onscroll (data) {
    setTimeout(apply_scroll_position, 200)
    function apply_scroll_position () {
      if (entries) {
        entries.scrollLeft = data
      }
    }
  }
}

function fallback_module () {
  return {
    api: fallback_instance,
    _: {
      DOCS: {
        $: ''
      },
      net_helper: {
        $: ''
      }
    }
  }
  function fallback_instance () {
    return {
      _: {
        DOCS: {
          0: ''
        },
        net_helper: {
          0: ''
        }
      },
      drive: {
        'icons/': {
          'cross.svg': {
            $ref: 'cross.svg'
          },
          '1.svg': {
            $ref: 'icon.svg'
          },
          '2.svg': {
            $ref: 'icon.svg'
          },
          '3.svg': {
            $ref: 'icon.svg'
          }
        },
        'actions/': {
          'commands.json': {
            raw: JSON.stringify([
              {
                name: 'New Tab',
                icon: 'plus',
                status: {
                  pinned: true,
                  default: true
                },
                steps: [
                  { name: 'Enter Tab Name', type: 'optional', is_completed: false, component: 'form_input', status: 'default', data: '' }
                ]
              },
              {
                name: 'Duplicate Tab',
                icon: 'copy',
                status: {
                  pinned: false,
                  default: false
                },
                steps: [
                  { name: 'Select Tab to Duplicate', type: 'mandatory', is_completed: false, component: 'form_input', status: 'default', data: '' },
                  { name: 'Enter New Tab Name', type: 'optional', is_completed: false, component: 'form_input', status: 'default', data: '' }
                ]
              },
              {
                name: 'Close Tab',
                icon: 'close',
                status: {
                  pinned: false,
                  default: true
                },
                steps: [
                  { name: 'Select Tab to Close', type: 'mandatory', is_completed: false, component: 'form_input', status: 'default', data: '' },
                  { name: 'Confirm Close', type: 'mandatory', is_completed: false, component: 'form_input', status: 'default', data: '' }
                ]
              }
            ])
          }
        },
        'variables/': {
          'tabs.json': {
            $ref: 'tabs.json'
          }
        },
        'scroll/': {
          'position.json': {
            raw: '100'
          }
        },
        'docs/': {
          'README.md': {
            $ref: 'README.md'
          }
        },
        'style/': {
          'theme.css': {
            $ref: 'style.css'
          }
        }
      }
    }
  }
}
