// program/program.js
const STATE = require('STATE')
const statedb = STATE(__filename)
const { get } = statedb(fallback_module)

const form_input = require('form_input')
const input_test = require('input_test')
const tile_split_choice = require('tile_split_choice')

program.form_input = form_input
program.input_test = input_test
program.tile_split_choice = tile_split_choice

module.exports = program

async function program (opts, protocol) {
  const { sdb } = await get(opts.sid)
  const { drive } = sdb
  const ids = opts.ids
  if (!ids || !ids.up) {
    throw new Error(`Component ${__filename} requires ids.up to be provided`)
  }
  // const by = id
  // const to = ids.up
  // const mid = 0
  // console.log('program-ids', by, to)
  const on = {
    style: inject,
    variables: onvariables
  }

  const _ = {
    up: null
  }

  if (protocol) {
    const send = protocol(onmessage)
    _.up = send
  }

  const el = document.createElement('div')
  const shadow = el.attachShadow({ mode: 'closed' })
  shadow.innerHTML = `
    <style></style>
  `
  const style = shadow.querySelector('style')

  await sdb.watch(onbatch)

  const parent_handler = {
    display_result,
    update_data
  }

  return el

  // --- Internal Functions ---
  async function onbatch (batch) {
    for (const { type, paths } of batch) {
      const data = await Promise.all(paths.map(load_path_raw))
      const func = on[type] || fail
      func(data, type)
    }

    function load_path_raw (path) { return drive.get(path).then(read_drive_file_raw) }
    function read_drive_file_raw (file) { return file.raw }
  }

  function fail (data, type) { console.warn('invalid message', { cause: { data, type } }) }

  function inject (data) {
    style.replaceChildren(create_style_element())

    function create_style_element () {
      const style_el = document.createElement('style')
      style_el.textContent = data[0]
      return style_el
    }
  }

  function onvariables (data) {
    // Dont get why we have this module.
  }

  function onmessage ({ type, data }) { 
    const handler = parent_handler[type](data, type) || fail
    handler(data, type)
  }
  function display_result (data) {
    console.log('Display Result:', data)
    alert(`Result of action(${data.selected_action ? data.selected_action : 'unknown'}): ${data.result ? data.result : 'no result'}`)
  }
  function update_data (data) { drive.put('variables/program.json', data) }
}

// --- Fallback Module ---
function fallback_module () {
  return {
    api: fallback_instance,
    _: {

      form_input: { $: '' },
      input_test: { $: '' },
      tile_split_choice: { $: '' }
    }
  }

  function fallback_instance () {
    return {
      drive: {
        'style/': {
          'program.css': {
            raw: `
              .main {
                display: flex;
                flex-direction: column;
                align-items: center;
              }
            `
          }
        },
        'variables/': {
          'program.json': { $ref: 'program.json' }
        }
      }
    }
  }
}
