// form_click_rate_test/form_click_rate_test.js
const STATE = require('STATE')
const statedb = STATE(__filename)
const { get } = statedb(fallback_module)
const DOCS = require('DOCS')
const net = require('net_helper')

module.exports = form_click_rate_test

async function form_click_rate_test (opts, invite) {
  const { id, sdb } = await get(opts.sid)
  const { drive } = sdb

  const on = {
    style: inject,
    data: ondata
  }

  let current_step = null
  let click_count = 0
  let start_time = 0
  let input_accessible = true
  const docs = DOCS(__filename)(opts.sid)
  const { io, _ } = net(id)
  docs.set_sys({ _, sdb, drive })
  const milliseconds_action = {
    name: 'Click Rate Result',
    info: 'Calculate and submit the milliseconds taken to complete 10 clicks.',
    icon: 'timer',
    status: {
      pinned: false,
      default: false
    },
    steps: []
  }

  const el = document.createElement('div')
  const shadow = el.attachShadow({ mode: 'closed' })
  shadow.innerHTML = `
  <div class="click-rate-test">
    <div class="title">Click Rate Test</div>
    <button class="click-btn" type="button">
      <span class="count">0</span>
      <span class="label">clicks</span>
    </button>
    <div class="hint">Click as fast as you can. The 10th click completes the action.</div>
    <div class="result"></div>
  </div>
  <style></style>
  `
  const style = shadow.querySelector('style')
  const button = shadow.querySelector('.click-btn')
  const count_el = shadow.querySelector('.count')
  const result_el = shadow.querySelector('.result')

  on_click_event_doc.info = 'Record one click in the 10-click sequence. This event is not an action until the 10th click.'
  const show_click_event_doc = docs.wrap(on_click_event_doc)
  button.onclick = on_click
  docs.register_actions([milliseconds_action])

  await sdb.watch(onbatch)

  const parent_handler = {
    step_data,
    reset_data
  }

  io.on = {
    up: io_up()
  }
  if (invite) io.accept(invite)

  return el

  function io_up () {
    return function onmessage ({ type, data }) {
      const handler = parent_handler[type] || fail
      handler(data, type)
    }
  }

  async function onbatch (batch) {
    for (const { type, paths } of batch) {
      const data = await Promise.all(paths.map(load_path_raw))
      const func = on[type] || fail
      func(data, type)
    }

    function load_path_raw (path) { return drive.get(path).then(read_drive_file_raw) }
    function read_drive_file_raw (file) { return file.raw }
  }

  function fail (data, type) { console.warn('invalid message', { cause: { data, type } }) }

  function inject (data) {
    style.replaceChildren(create_style_element())

    function create_style_element () {
      const style_el = document.createElement('style')
      style_el.textContent = data[0]
      return style_el
    }
  }

  function ondata (data) {
    if (data.length === 0 || !data[0]) return
    click_count = data[0].click_count || 0
    update_count()
    if (data[0].result) show_result(data[0].result)
  }

  function step_data (data) {
    current_step = data
    input_accessible = data.is_accessible !== false
    button.disabled = !input_accessible
    reset_clicks()
  }

  function reset_data () {
    reset_clicks()
    drive.put('data/form_click_rate_test.json', {
      click_count: 0,
      result: ''
    })
  }

  function on_click_event_doc () {}

  async function on_click (event) {
    if (!input_accessible) return
    if (click_count === 0) start_time = Date.now()

    click_count += 1
    update_count()

    if (click_count < 10) {
      await show_click_event_doc(event)
      await drive.put('data/form_click_rate_test.json', { click_count, result: '' })
      _.up('action_incomplete', {}, { value: click_count, index: get_step_index() })
      return
    }

    await run_milliseconds_action()
  }

  async function run_milliseconds_action () {
    if (docs.show_action_info(milliseconds_action)) return

    const result = '10 clicks in ' + get_elapsed_time() + ' ms'
    show_result(result)
    await drive.put('data/form_click_rate_test.json', { click_count, result })
    _.up('action_submitted', {}, { value: result, index: get_step_index() })
    _.up('action_complete', {}, { value: result })
  }

  function get_elapsed_time () { return Date.now() - start_time }

  function get_step_index () {
    if (current_step && current_step.index !== undefined) return current_step.index
    return 0
  }

  function reset_clicks () {
    click_count = 0
    start_time = 0
    result_el.textContent = ''
    update_count()
  }

  function update_count () { count_el.textContent = click_count }
  function show_result (result) { result_el.textContent = result }
}

function fallback_module () {
  return {
    api: fallback_instance,
    _: {
      DOCS: { $: '' },
      net_helper: { $: '' }
    }
  }

  function fallback_instance () {
    return {
      _: {
        DOCS: { 0: '' },
        net_helper: { 0: '' }
      },
      drive: {
        'style/': {
          'form_click_rate_test.css': { $ref: 'form_click_rate_test.css' }
        },
        'data/': {
          'form_click_rate_test.json': { raw: { click_count: 0, result: '' } }
        },
        'docs/': { 'README.md': { raw: '# Click Rate Test\nClick 10 times as fast as possible.' } }
      }
    }
  }
}
