// taskbar/taskbar.js
const STATE = require('STATE')
const statedb = STATE(__filename)
const { get } = statedb(fallback_module)
const manager = require('manager')
const exec = require('exec')
const tabsbar = require('tabsbar')

module.exports = taskbar

async function taskbar (opts, protocol) {
  const { id, sdb } = await get(opts.sid)
  const { drive } = sdb
  const ids = opts.ids
  if (!ids || !ids.up) {
    throw new Error(`Component ${__filename} requires ids.up to be provided`)
  }
  // const by = id
  // const to = ids.up

  const on = {
    style: inject
  }
  const on_message = {
    update_steps_wizard_for_app: handle_update_steps_wizard_for_app
  }

  const el = document.createElement('div')
  const shadow = el.attachShadow({ mode: 'closed' })

  shadow.innerHTML = `
  <div class="taskbar-container main">
    <div class="exec-slot"></div>
    <div class="bottom-slot">
      <div class="manager-slot"></div>
      <div class="tabsbar-slot"></div>
    </div>
  </div>
  <style>
  </style>`
  const style = shadow.querySelector('style')
  const exec_slot = shadow.querySelector('.exec-slot')
  const manager_slot = shadow.querySelector('.manager-slot')
  const tabsbar_slot = shadow.querySelector('.tabsbar-slot')

  const subs = await sdb.watch(onbatch)
  let send = null
  let _ = null
  if (protocol) {
    send = protocol(msg => onmessage(msg))
    _ = { up: send, manager: null, exec: null, tabsbar: null }
  }
  const manager_el = protocol ? await manager({ ...subs[0], ids: { up: id } }, manager_protocol) : await manager({ ...subs[0], ids: { up: id } })
  manager_el.classList.add('replaced-manager')
  manager_slot.replaceWith(manager_el)

  const exec_el = protocol ? await exec({ ...subs[1], ids: { up: id } }, exec_protocol) : await exec({ ...subs[1], ids: { up: id } })
  exec_el.classList.add('replaced-exec')
  exec_slot.replaceWith(exec_el)

  const tabsbar_el = protocol ? await tabsbar({ ...subs[2], ids: { up: id } }, tabsbar_protocol) : await tabsbar({ ...subs[2], ids: { up: id } })
  tabsbar_el.classList.add('replaced-tabsbar')
  tabsbar_slot.replaceWith(tabsbar_el)

  return el

  async function onbatch (batch) {
    for (const { type, paths } of batch) {
      const data = await Promise.all(paths.map(path => drive.get(path).then(file => file.raw)))
      const func = on[type] || fail
      func(data, type)
    }
  }

  function fail (data, type) { console.warn('invalid message', { cause: { data, type } }) }

  function inject (data) {
    style.innerHTML = data.join('\n')
  }

  // ---------
  // PROTOCOLS
  // ---------
  function manager_protocol (send) {
    _.manager = send
    const action_handlers = {
      update_data: manager__forward_exec,
      form_data: manager__forward_exec,
      clean_up: manager__forward_exec,
      action_submitted: manager__forward_exec,
      activate_steps_wizard: manager__forward_exec,
      render_form: manager__forward_exec,
      selected_action: manager__forward_exec
    }
    return on
    function on (msg) {
      const handler = action_handlers[msg.type] || manager__forward_up
      handler(msg)
    }

    function manager__forward_exec (msg) {
      _.exec && _.exec(msg)
    }

    function manager__forward_up (msg) {
      _.up && _.up(msg)
    }
  }

  function exec_protocol (send) {
    _.exec = send
    const action_handlers = {
      load_actions: exec__forward_manager,
      step_clicked: exec__forward_manager,
      show_submit_btn: exec__forward_manager,
      hide_submit_btn: exec__forward_manager
    }
    return on
    function on (msg) {
      const handler = action_handlers[msg.type] || exec__noop
      handler(msg)
      _.up && _.up(msg)
    }

    function exec__forward_manager (msg) {
      _.manager && _.manager(msg)
    }

    function exec__noop () {}
  }

  function tabsbar_protocol (send) {
    _.tabsbar = send
    const action_handlers = {
      docs_toggle: tabsbar__docs_toggle
    }
    return on
    function on (msg) {
      const handler = action_handlers[msg.type] || tabsbar__noop
      handler(msg)
      _.up && _.up(msg)
    }

    function tabsbar__docs_toggle (msg) {
      _.manager && _.manager(msg)
      _.exec && _.exec(msg)
    }

    function tabsbar__noop () {}
  }

  function onmessage (msg) {
    const handler = on_message[msg.type] || onmessage_forward_manager
    handler(msg)
  }

  function handle_update_steps_wizard_for_app (msg) {
    _.exec && _.exec(msg)
  }

  function onmessage_forward_manager (msg) {
    _.manager && _.manager(msg)
  }
}

function fallback_module () {
  return {
    api: fallback_instance,
    _: {
      manager: {
        $: ''
      },
      exec: {
        $: ''
      },
      tabsbar: {
        $: ''
      }
    }
  }

  function fallback_instance () {
    return {
      _: {
        manager: {
          0: '',
          mapping: {
            icons: 'icons',
            style: 'style',
            variables: 'variables',
            data: 'data',
            actions: 'actions',
            hardcons: 'hardcons',
            prefs: 'prefs',
            docs: 'docs'
          }
        },
        exec: {
          0: '',
          mapping: {
            style: 'style',
            variables: 'variables',
            docs: 'docs',
            data: 'data'
          }
        },
        tabsbar: {
          0: '',
          mapping: {
            icons: 'icons',
            style: 'style',
            docs: 'docs',
            actions: 'actions'
          }
        }
      },
      drive: {
        'style/': {
          'theme.css': {
            raw: `
              .taskbar-container {
                display: flex;
                background: #2d2d2d;
                column-gap: 1px;
                flex-direction: column;
                align-content: center;
                justify-content: center;
                container-type: inline-size;
              }
              .replaced-tabsbar {
                display: flex;
                flex: auto;
              }
              .replaced-manager {
                display: flex;
              }
              .replaced-exec {
                display: flex;
              }
              .bottom-slot {
                display: flex;
                flex-direction: row;
                justify-content: space-between;
              }
              @container (max-width: 700px) {
                .bottom-slot {
                  flex-direction: column;
                }
              }
            `
          }
        },
        'icons/': {},
        'variables/': {},
        'data/': {},
        'actions/': {},
        'hardcons/': {},
        'prefs/': {},
        'docs/': {
          'README.md': {
            $ref: 'README.md'
          }
        }
      }
    }
  }
}
