// focus_tracker/focus_tracker.js
const STATE = require('STATE')
const statedb = STATE(__filename)
const { get } = statedb(fallback_module)

module.exports = focus_tracker

async function focus_tracker (opts, protocol) {
  const { id, sdb } = await get(opts.sid)
  const { drive } = sdb
  const ids = opts.ids
  if (!ids || !ids.up) {
    throw new Error(`Component ${__filename} requires ids.up to be provided`)
  }
  const by = id
  const to = ids.up

  const on = {
    focused
  }
  const on_message = {
    ui_focus: handle_ui_focus
  }
  // Keep track of the last focused element
  let last_focused = null
  let mid = 0
  let _ = null

  if (protocol) {
    const send = protocol(onmessage)
    _ = { up: send }
  }

  function onmessage (msg) {
    const handler = on_message[msg.type] || onmessage_fail
    handler(msg)
  }

  function handle_ui_focus (msg) {
    if (last_focused !== msg.data.type) {
      const head = [by, to, mid++]
      const refs = {}
      _.up({ head, refs, type: 'focused_app_changed', data: msg.data })
    }
    drive.put('focused/current.json', { value: msg.data.type })
  }

  function onmessage_fail (msg) { fail(msg.data, msg.type) }

  await sdb.watch(onbatch)

  async function onbatch (batch) {
    for (const { type, paths } of batch) {
      const data = await Promise.all(paths.map(load_path_raw))
      const func = on[type] || fail
      func(data, type)
    }

    function load_path_raw (path) { return drive.get(path).then(read_drive_file_raw) }
    function read_drive_file_raw (file) { return file.raw }
  }
  function fail (data, type) { console.warn('invalid message', { cause: { data, type } }) }
  function focused (data) {
    const tmp = typeof data[0] === 'string' ? JSON.parse(data[0]) : data[0]
    last_focused = tmp.value
  }
}

function fallback_module () {
  return {
    api: fallback_instance
  }
  function fallback_instance () {
    return {
      drive: {
        'focused/': {
          'current.json': {
            raw: { value: 'default' }
          }
        },
        'docs/': {
          'README.md': {
            $ref: 'README.md'
          }
        }
      }
    }
  }
}
